diff -ruN expat-2.1.0/lib/xmlparse.c expat-2.1.0.aros/lib/xmlparse.c
--- expat-2.1.0/lib/xmlparse.c	2012-03-11 05:13:12.000000000 +0000
+++ expat-2.1.0.aros/lib/xmlparse.c	2023-05-17 10:41:46.568843700 +0000
@@ -16,6 +16,10 @@
 #include "macconfig.h"
 #elif defined(__amigaos__)
 #include "amigaconfig.h"
+#elif defined(__AROS__)
+#include "arosconfig.h"
+#undef assert
+#define assert(x) ASSERT(x)
 #elif defined(__WATCOMC__)
 #include "watcomconfig.h"
 #elif defined(HAVE_EXPAT_CONFIG_H)
diff -ruN expat-2.1.0/lib/xmlrole.c expat-2.1.0.aros/lib/xmlrole.c
--- expat-2.1.0/lib/xmlrole.c	2009-01-31 06:21:09.000000000 +0000
+++ expat-2.1.0.aros/lib/xmlrole.c	2023-05-17 10:41:46.569344400 +0000
@@ -10,6 +10,8 @@
 #include "macconfig.h"
 #elif defined(__amigaos__)
 #include "amigaconfig.h"
+#elif defined(__AROS__)
+#include "arosconfig.h"
 #elif defined(__WATCOMC__)
 #include "watcomconfig.h"
 #else
diff -ruN expat-2.1.0/lib/xmltok.c expat-2.1.0.aros/lib/xmltok.c
--- expat-2.1.0/lib/xmltok.c	2009-01-31 06:21:09.000000000 +0000
+++ expat-2.1.0.aros/lib/xmltok.c	2023-05-17 10:41:46.570346300 +0000
@@ -10,6 +10,8 @@
 #include "macconfig.h"
 #elif defined(__amigaos__)
 #include "amigaconfig.h"
+#elif defined(__AROS__)
+#include "arosconfig.h"
 #elif defined(__WATCOMC__)
 #include "watcomconfig.h"
 #else
