diff -ur /data/deadwood/source/Mesa-7.9-staging/src/gallium/auxiliary/os/os_time.c ./src/gallium/auxiliary/os/os_time.c
--- /data/deadwood/source/Mesa-7.9-staging/src/gallium/auxiliary/os/os_time.c	2010-10-02 00:51:28.000000000 +0200
+++ ./src/gallium/auxiliary/os/os_time.c	2010-12-26 19:20:30.000000000 +0100
@@ -37,7 +37,7 @@
 
 #if !defined(PIPE_OS_EMBEDDED)
 
-#if defined(PIPE_OS_LINUX) || defined(PIPE_OS_BSD) || defined(PIPE_OS_SOLARIS) || defined(PIPE_OS_APPLE) || defined(PIPE_OS_HAIKU) || defined(PIPE_OS_CYGWIN)
+#if defined(PIPE_OS_LINUX) || defined(PIPE_OS_BSD) || defined(PIPE_OS_SOLARIS) || defined(PIPE_OS_APPLE) || defined(PIPE_OS_HAIKU) || defined(PIPE_OS_CYGWIN) || defined(PIPE_OS_AROS)
 #  include <sys/time.h> /* timeval */
 #elif defined(PIPE_SUBSYSTEM_WINDOWS_DISPLAY)
 #  include <windows.h>
@@ -57,7 +57,7 @@
 int64_t
 os_time_get(void)
 {
-#if defined(PIPE_OS_UNIX)
+#if defined(PIPE_OS_UNIX) || defined(PIPE_OS_AROS)
 
    struct timeval tv;
    gettimeofday(&tv, NULL);
diff -ur /data/deadwood/source/Mesa-7.9-staging/src/gallium/auxiliary/util/u_atomic.h ./src/gallium/auxiliary/util/u_atomic.h
--- /data/deadwood/source/Mesa-7.9-staging/src/gallium/auxiliary/util/u_atomic.h	2010-10-02 01:55:56.000000000 +0200
+++ ./src/gallium/auxiliary/util/u_atomic.h	2010-12-26 19:20:30.000000000 +0100
@@ -29,8 +29,12 @@
 #define PIPE_ATOMIC_ASM_MSVC_X86                
 #elif (defined(PIPE_CC_GCC) && defined(PIPE_ARCH_X86))
 #define PIPE_ATOMIC_ASM_GCC_X86
+#elif (defined(PIPE_CC_GCC) && defined(PIPE_ARCH_ARM))
+#define PIPE_ATOMIC_ASM_GCC_ARM
 #elif (defined(PIPE_CC_GCC) && defined(PIPE_ARCH_X86_64))
 #define PIPE_ATOMIC_ASM_GCC_X86_64
+#elif defined(PIPE_OS_AROS) && defined(PIPE_ARCH_M68K)
+#define PIPE_ATOMIC_OS_AROS_CPU_M68K
 #elif defined(PIPE_CC_GCC) && (PIPE_CC_GCC_VERSION >= 401)
 #define PIPE_ATOMIC_GCC_INTRINSIC
 #else
@@ -130,6 +134,69 @@
 
 #endif
 
+#if defined(PIPE_ATOMIC_ASM_GCC_ARM)
+
+#define PIPE_ATOMIC "GCC ARM assembly"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#define p_atomic_set(_v, _i) (*(_v) = (_i))
+#define p_atomic_read(_v) (*(_v))
+
+static INLINE boolean
+p_atomic_dec_zero(int32_t *v)
+{
+   unsigned long temp;
+   int result;
+   unsigned long cc;
+   __asm__ __volatile__("\n1: ldrex %0, [%3]; subs %0, %0, #1; moveq %2, #1; movne %2, #0; strex %1, %0, [%3]; teq %1, #0; bne 1b"
+						   :"=&r"(result), "=&r"(temp), "=&r"(cc)
+						   :"r"(v)
+						   :"cc");
+   return cc;
+}
+
+static INLINE void
+p_atomic_inc(int32_t *v)
+{
+   unsigned long temp;
+   int result;
+   __asm__ __volatile__("\n1: ldrex %0, [%2]; add %0, %0, #1; strex %1, %0, [%2]; teq %1, #0; bne 1b"
+		   	   	   	   	   :"=&r"(result), "=&r"(temp)
+		   	   	   	   	   :"r"(v)
+		   	   	   	   	   :"cc");
+}
+
+static INLINE void
+p_atomic_dec(int32_t *v)
+{
+   unsigned long temp;
+   int result;
+   __asm__ __volatile__("\n1: ldrex %0, [%2]; sub %0, %0, #1; strex %1, %0, [%2]; teq %1, #0; bne 1b"
+			   	   	   	   :"=&r"(result), "=&r"(temp)
+			   	   	   	   :"r"(v)
+			   	   	   	   :"cc");
+}
+
+static INLINE int32_t
+p_atomic_cmpxchg(int32_t *v, int32_t old, int32_t _new)
+{
+	  int32_t oldval;
+	   unsigned long temp;
+	   __asm__ __volatile__("\n1: ldrex %0,[%2]; teq %0, %3; strexeq %1, %4, [%2]; teq %1, #0; bne 1b"
+	                                                   :"=&r"(oldval), "=&r"(temp)
+	                                                   :"r"(v), "Ir"(old), "r"(_new)
+	                                                   :"cc");
+	   return oldval;
+}
+
+#ifdef __cplusplus
+}
+#endif
+
+#endif
 
 
 /* Implementation using GCC-provided synchronization intrinsics
@@ -343,6 +410,64 @@
 #endif
 
 
+#if defined(PIPE_ATOMIC_OS_AROS_CPU_M68K)
+
+#define PIPE_ATOMIC "AROS OS atomic functions"
+
+#include <aros/atomic.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#define p_atomic_set(_v, _i) (*(_v) = (_i))
+#define p_atomic_read(_v) (*(_v))
+
+static INLINE boolean
+p_atomic_dec_zero(int32_t *v)
+{
+   boolean n;
+ 
+   /* FIXME: AROS needs an atomic decrement and return... */
+   Disable();
+   AROS_ATOMIC_DEC(*(LONG *)v);
+   n = (*v != 0) ? TRUE : FALSE;
+   Enable();
+
+   return n;
+}
+
+#define p_atomic_inc(_v) AROS_ATOMIC_INC(*(LONG *)_v)
+#define p_atomic_dec(_v) AROS_ATOMIC_DEC(*(LONG *)_v)
+
+static INLINE int32_t
+p_atomic_cmpxchg(int32_t *v, int32_t o, int32_t n)
+{
+	int32_t ret;
+
+	/* FIXME: AROS needs an atomic cmpxchg, using CAS.
+	 * However we can't do this if:
+	 *  a) We are on a 68000 or
+	 *  b) The 'v' points to Chip RAM (no r/m/w possible)
+	 *
+	 *  Settle for Disable()/Enable() for now.
+	 */
+	Disable();
+	if (*v == o)
+		*v = (n);
+	ret = *v;
+	Enable();
+
+	return ret;
+}
+
+#ifdef __cplusplus
+}
+#endif
+
+#endif
+
+
 #ifndef PIPE_ATOMIC
 #error "No pipe_atomic implementation selected"
 #endif
diff -ur /data/deadwood/source/Mesa-7.9-staging/src/gallium/auxiliary/util/u_cpu_detect.c ./src/gallium/auxiliary/util/u_cpu_detect.c
--- /data/deadwood/source/Mesa-7.9-staging/src/gallium/auxiliary/util/u_cpu_detect.c	2010-10-02 00:51:28.000000000 +0200
+++ ./src/gallium/auxiliary/util/u_cpu_detect.c	2010-12-26 19:20:30.000000000 +0100
@@ -195,9 +195,12 @@
    );
 #elif defined(PIPE_CC_GCC) && defined(PIPE_ARCH_X86_64)
    __asm __volatile (
+     "pushq %%rbx\n\t"
      "cpuid\n\t"
+     "movl %%ebx, %%esi\n\t"
+     "popq %%rbx"
      : "=a" (p[0]),
-       "=b" (p[1]),
+       "=S" (p[1]),
        "=c" (p[2]),
        "=d" (p[3])
      : "0" (ax)
diff -ur /data/deadwood/source/Mesa-7.9-staging/src/gallium/auxiliary/util/u_math.c ./src/gallium/auxiliary/util/u_math.c
--- /data/deadwood/source/Mesa-7.9-staging/src/gallium/auxiliary/util/u_math.c	2010-02-05 01:10:39.000000000 +0100
+++ ./src/gallium/auxiliary/util/u_math.c	2010-12-26 19:20:30.000000000 +0100
@@ -42,6 +42,13 @@
       pow2_table[i] = (float) pow(2.0, (i - POW2_TABLE_OFFSET) / POW2_TABLE_SCALE);
 }
 
+#if defined(PIPE_OS_AROS)
+/*
+ * NOTE: log_base_2(x) = log(x) / log(2)
+ * NOTE: 1.442695 = 1/log(2).
+ */
+#define log2(x)  ((float) (log(x) * 1.442695f))
+#endif
 
 /** log2(x), for x in [1.0, 2.0) */
 float log2_table[LOG2_TABLE_SIZE];
diff -ur /data/deadwood/source/Mesa-7.9-staging/src/gallium/auxiliary/util/u_math.h ./src/gallium/auxiliary/util/u_math.h
--- /data/deadwood/source/Mesa-7.9-staging/src/gallium/auxiliary/util/u_math.h	2010-10-02 00:51:28.000000000 +0200
+++ ./src/gallium/auxiliary/util/u_math.h	2010-12-21 19:49:33.000000000 +0100
@@ -394,6 +394,8 @@
 }
 #elif defined(__MINGW32__)
 #define ffs __builtin_ffs
+#elif defined(PIPE_OS_AROS)
+#define ffs __builtin_ffs
 #endif
 
 #ifdef __MINGW32__
@@ -498,7 +500,13 @@
 static INLINE unsigned
 util_bitcount(unsigned n)
 {
-#if defined(PIPE_CC_GCC)
+#if defined(PIPE_OS_AROS)
+   unsigned int bits;
+   for (bits = 0; n > 0; n = n >> 1) {
+      bits += (n & 1);
+   }
+   return bits;
+#elif defined(PIPE_CC_GCC)
    return __builtin_popcount(n);
 #else
    /* K&R classic bitcount.
diff -ur /data/deadwood/source/Mesa-7.9-staging/src/gallium/drivers/nvfx/nvfx_fragprog.c ./src/gallium/drivers/nvfx/nvfx_fragprog.c
--- /data/deadwood/source/Mesa-7.9-staging/src/gallium/drivers/nvfx/nvfx_fragprog.c	2010-10-05 03:58:00.000000000 +0200
+++ ./src/gallium/drivers/nvfx/nvfx_fragprog.c	2010-12-26 19:20:30.000000000 +0100
@@ -1556,7 +1556,7 @@
 			struct nvfx_fragment_program_bo* next = fpbo->next;
 			nouveau_bo_unmap(fpbo->bo);
 			nouveau_bo_ref(0, &fpbo->bo);
-			free(fpbo);
+			os_free_aligned(fpbo);
 			fpbo = next;
 		}
 		while(fpbo != fp->fpbo);
diff -ur /data/deadwood/source/Mesa-7.9-staging/src/gallium/include/pipe/p_config.h ./src/gallium/include/pipe/p_config.h
--- /data/deadwood/source/Mesa-7.9-staging/src/gallium/include/pipe/p_config.h	2010-10-02 00:51:28.000000000 +0200
+++ ./src/gallium/include/pipe/p_config.h	2010-12-26 19:20:30.000000000 +0100
@@ -82,6 +82,10 @@
 #define PIPE_ARCH_X86
 #endif
 
+#if defined(__arm__)
+#define PIPE_ARCH_ARM
+#endif
+
 #if defined(__x86_64__) /* gcc */ || defined(_M_X64) /* msvc */ || defined(_M_AMD64) /* msvc */ || defined(__x86_64) /* Sun cc */
 #define PIPE_ARCH_X86_64
 #endif
@@ -106,12 +110,16 @@
 #endif
 #endif
 
+#if defined(__mc68000) /* gcc */
+#define PIPE_ARCH_M68K
+#endif
+
 
 /*
  * Endian detection.
  */
 
-#if defined(PIPE_ARCH_X86) || defined(PIPE_ARCH_X86_64)
+#if defined(PIPE_ARCH_X86) || defined(PIPE_ARCH_X86_64) || defined(PIPE_ARCH_ARM)
 #define PIPE_ARCH_LITTLE_ENDIAN
 #elif defined(PIPE_ARCH_PPC) || defined(PIPE_ARCH_PPC_64)
 #define PIPE_ARCH_BIG_ENDIAN
@@ -128,6 +136,15 @@
  * See subsystem below for a more fine-grained distinction.
  */
 
+/*
+ * In some situations __AROS__ definition can coexist with anything of
+ * the below (for example this happens on ARM port which is compiled using Linux
+ * compiler at the moment). So if we have __AROS__ we don't evaluate anything else
+ */
+#if defined(__AROS__)
+#define PIPE_OS_AROS
+#else
+
 #if defined(__linux__)
 #define PIPE_OS_LINUX
 #define PIPE_OS_UNIX
@@ -180,6 +197,8 @@
 #define PIPE_OS_UNIX
 #endif
 
+#endif
+
 /*
  * Try to auto-detect the subsystem.
  * 
diff -ur /data/deadwood/source/Mesa-7.9-staging/src/glsl/ir.h ./src/glsl/ir.h
--- /data/deadwood/source/Mesa-7.9-staging/src/glsl/ir.h	2010-10-02 00:51:28.000000000 +0200
+++ ./src/glsl/ir.h	2010-12-26 19:20:30.000000000 +0100
@@ -41,6 +41,14 @@
 #define ARRAY_SIZE(x) (sizeof(x) / sizeof(x[0]))
 #endif
 
+#if defined(__AROS__)
+#include <aros/debug.h>
+#undef VOLATILE
+#undef STATIC
+#define log2(x)             ((float) (log(x) * 1.442695f))
+#define printf(fmt, ...)    bug(fmt, ##__VA_ARGS__)
+#endif
+
 /**
  * \defgroup IR Intermediate representation nodes
  *
diff -ur /data/deadwood/source/Mesa-7.9-staging/src/glsl/s_expression.cpp ./src/glsl/s_expression.cpp
--- /data/deadwood/source/Mesa-7.9-staging/src/glsl/s_expression.cpp	2010-10-02 00:51:28.000000000 +0200
+++ ./src/glsl/s_expression.cpp	2010-12-26 19:20:30.000000000 +0100
@@ -28,6 +28,11 @@
 #include <assert.h>
 #include "s_expression.h"
 
+#if defined(__AROS__)
+#include <aros/debug.h>
+#define printf(fmt, ...)    bug(fmt, ##__VA_ARGS__)
+#endif
+
 s_symbol::s_symbol(const char *tmp, size_t n)
 {
    this->str = talloc_strndup (this, tmp, n);
diff -ur /data/deadwood/source/Mesa-7.9-staging/src/mapi/glapi/glapi.h ./src/mapi/glapi/glapi.h
--- /data/deadwood/source/Mesa-7.9-staging/src/mapi/glapi/glapi.h	2010-10-02 00:51:28.000000000 +0200
+++ ./src/mapi/glapi/glapi.h	2010-12-26 19:20:30.000000000 +0100
@@ -82,6 +82,32 @@
 typedef void (*_glapi_proc)(void);
 struct _glapi_table;
 
+#if defined(__AROS__)
+
+#if defined(__i386__)
+register struct MesaBase * REGMesaBase __asm__("ebx");
+#elif defined(__x86_64__)
+register struct MesaBase * REGMesaBase __asm__("rbx");
+#elif defined(__arm__)
+register struct MesaBase * REGMesaBase __asm__("r10");
+#elif defined(PPC) || defined (__powerpc__)
+/*
+ * r11 or r12 emit call clobbered register warnings
+ * r13 works as well, but is small data area pointer
+ * in SysV ABI, which we might use in the future
+ */
+register struct MesaBase * REGMesaBase __asm__("r14");
+#else
+#error Select register for your architecture
+#endif
+
+#define SAVE_REG            struct MesaBase * reg = REGMesaBase;
+#define PUT_MESABASE_IN_REG REGMesaBase = (struct MesaBase *)MesaBase;
+#define RESTORE_REG         REGMesaBase = reg;
+extern void                     **GETMESABASECTX(void);
+extern struct _glapi_table      **GETMESABASEDDISPATCH(void);
+#endif
+
 
 #if defined (GLX_USE_TLS)
 
@@ -99,8 +125,18 @@
 
 #else
 
+#if defined(__AROS__)
+# if defined(USE_MGL_NAMESPACE)
+#  define _mglapi_Context *(GETMESABASECTX())
+#  define _mglapi_Dispatch *(GETMESABASEDDISPATCH())
+# else
+#  define _glapi_Context *(GETMESABASECTX())
+#  define _glapi_Dispatch *(GETMESABASEDDISPATCH())
+# endif
+#else
 _GLAPI_EXPORT extern struct _glapi_table *_glapi_Dispatch;
 _GLAPI_EXPORT extern void *_glapi_Context;
+#endif
 
 # ifdef THREADS
 
diff -ur /data/deadwood/source/Mesa-7.9-staging/src/mapi/mapi/u_current.c ./src/mapi/mapi/u_current.c
--- /data/deadwood/source/Mesa-7.9-staging/src/mapi/mapi/u_current.c	2010-10-02 00:51:28.000000000 +0200
+++ ./src/mapi/mapi/u_current.c	2010-12-27 11:22:46.000000000 +0100
@@ -111,9 +111,22 @@
 
 #else
 
+#if defined(__AROS__)
+#include "aros/arosmesa_intern.h"
+void **GETMESABASECTX(void)
+{
+    return &(REGMesaBase->mglb_CurrentContext);
+}
+
+struct _glapi_table **GETMESABASEDDISPATCH(void)
+{
+    return (struct _glapi_table **)&(REGMesaBase->mglb_Dispatch);
+}
+#else
 struct mapi_table *u_current_table =
    (struct mapi_table *) table_noop_array;
 void *u_current_user;
+#endif
 
 #ifdef THREADS
 struct u_tsd u_current_table_tsd;
diff -ur /data/deadwood/source/Mesa-7.9-staging/src/mapi/mapi/u_thread.c ./src/mapi/mapi/u_thread.c
--- /data/deadwood/source/Mesa-7.9-staging/src/mapi/mapi/u_thread.c	2010-10-02 00:51:28.000000000 +0200
+++ ./src/mapi/mapi/u_thread.c	2010-12-26 19:20:30.000000000 +0100
@@ -222,7 +222,7 @@
  */
 
 unsigned long
-_glthread_GetID(void)
+u_thread_self(void)
 {
    return 0;
 }
diff -ur /data/deadwood/source/Mesa-7.9-staging/src/mesa/main/compiler.h ./src/mesa/main/compiler.h
--- /data/deadwood/source/Mesa-7.9-staging/src/mesa/main/compiler.h	2010-10-02 00:51:28.000000000 +0200
+++ ./src/mesa/main/compiler.h	2010-12-26 19:20:30.000000000 +0100
@@ -251,6 +251,8 @@
 #elif defined(__APPLE__)
 #include <CoreFoundation/CFByteOrder.h>
 #define CPU_TO_LE32( x )	CFSwapInt32HostToLittle( x )
+#elif defined(__AROS__)
+#define CPU_TO_LE32( x )    AROS_BE2LONG( x )
 #elif (defined(_AIX) || defined(__blrts))
 static INLINE GLuint CPU_TO_LE32(GLuint x)
 {
@@ -318,6 +320,9 @@
 /**
  * ASSERT macro
  */
+#if defined(__AROS__)
+#undef ASSERT
+#endif
 #if !defined(_WIN32_WCE)
 #if defined(BUILD_FOR_SNAP) && defined(CHECKED)
 #  define ASSERT(X)   _CHECK(X) 
diff -ur /data/deadwood/source/Mesa-7.9-staging/src/mesa/main/context.c ./src/mesa/main/context.c
--- /data/deadwood/source/Mesa-7.9-staging/src/mesa/main/context.c	2010-10-05 03:58:00.000000000 +0200
+++ ./src/mesa/main/context.c	2010-12-26 19:20:30.000000000 +0100
@@ -439,7 +439,7 @@
    /* Hopefully atexit() is widely available.  If not, we may need some
     * #ifdef tests here.
     */
-   atexit(_mesa_destroy_shader_compiler);
+//TODO FIXME   atexit(_mesa_destroy_shader_compiler);
 
    dummy_enum_func();
 }
diff -ur /data/deadwood/source/Mesa-7.9-staging/src/mesa/main/imports.c ./src/mesa/main/imports.c
--- /data/deadwood/source/Mesa-7.9-staging/src/mesa/main/imports.c	2010-10-02 00:51:28.000000000 +0200
+++ ./src/mesa/main/imports.c	2010-12-26 19:20:30.000000000 +0100
@@ -457,7 +457,7 @@
 int
 _mesa_ffs(int32_t i)
 {
-#if (defined(_WIN32) ) || defined(__IBMC__) || defined(__IBMCPP__)
+#if (defined(_WIN32) ) || defined(__IBMC__) || defined(__IBMCPP__) || defined(__AROS__)
    register int bit = 0;
    if (i != 0) {
       if ((i & 0xffff) == 0) {
@@ -520,7 +520,8 @@
 unsigned int
 _mesa_bitcount(unsigned int n)
 {
-#if defined(__GNUC__) && \
+/* AROS: don't use __builtin_popcount */
+#if !defined(__AROS__) && defined(__GNUC__) && \
 	((_GNUC__ == 3 && __GNUC_MINOR__ >= 4) || __GNUC__ >= 4)
    return __builtin_popcount(n);
 #else
diff -ur /data/deadwood/source/Mesa-7.9-staging/src/mesa/main/querymatrix.c ./src/mesa/main/querymatrix.c
--- /data/deadwood/source/Mesa-7.9-staging/src/mesa/main/querymatrix.c	2010-10-05 03:58:00.000000000 +0200
+++ ./src/mesa/main/querymatrix.c	2010-12-26 19:20:30.000000000 +0100
@@ -76,6 +76,10 @@
      (defined(__sun) && defined(__GNUC__))
 
 /* fpclassify is available. */
+#elif defined(__AROS__)
+
+#undef fpclassify
+#define fpclassify(x)   FP_NORMAL
 
 #elif !defined(_XOPEN_SOURCE) || _XOPEN_SOURCE < 600
 
