diff -ruN libunwind-11.0.0.src/src/config.h libunwind-11.0.0.src.aros/src/config.h
--- libunwind-11.0.0.src/src/config.h	2020-10-07 10:10:48.000000000 +0000
+++ libunwind-11.0.0.src.aros/src/config.h	2025-07-11 14:43:23.899956677 +0000
@@ -36,6 +36,12 @@
     #define _LIBUNWIND_SUPPORT_COMPACT_UNWIND
     #define _LIBUNWIND_SUPPORT_DWARF_UNWIND   1
   #endif
+#elif defined(__AROS__)
+  #define _LIBUNWIND_IS_BAREMETAL
+  #define _LIBUNWIND_SUPPORT_DWARF_UNWIND   1
+  #ifdef _LIBUNWIND_USE_DLADDR
+  #undef _LIBUNWIND_USE_DLADDR
+  #endif
 #elif defined(_WIN32)
   #ifdef __SEH__
     #define _LIBUNWIND_SUPPORT_SEH_UNWIND 1
