diff -ruN llvm-11.0.0.src/include/llvm/ADT/Triple.h llvm-11.0.0.src.aros/include/llvm/ADT/Triple.h
--- llvm-11.0.0.src/include/llvm/ADT/Triple.h	2020-10-07 10:10:48.000000000 +0000
+++ llvm-11.0.0.src.aros/include/llvm/ADT/Triple.h	2020-10-07 10:10:48.000000000 +0000
@@ -161,6 +161,7 @@
     UnknownOS,
 
     Ananas,
+	AROS,
     CloudABI,
     Darwin,
     DragonFly,
@@ -448,6 +449,11 @@
     return getOS() == Triple::Darwin || getOS() == Triple::MacOSX;
   }
 
+  /// Is this an AROS triple.
+  bool isAROS() const {
+    return getOS() == Triple::AROS;
+  }
+
   /// Is this an iOS triple.
   /// Note: This identifies tvOS as a variant of iOS. If that ever
   /// changes, i.e., if the two operating systems diverge or their version
diff -ruN llvm-11.0.0.src/include/llvm/Support/Signals.h llvm-11.0.0.src.aros/include/llvm/Support/Signals.h
--- llvm-11.0.0.src/include/llvm/Support/Signals.h	2020-10-07 10:10:48.000000000 +0000
+++ llvm-11.0.0.src.aros/include/llvm/Support/Signals.h	2025-06-30 02:39:23.759636178 +0000
@@ -15,6 +15,7 @@
 #define LLVM_SUPPORT_SIGNALS_H
 
 #include <string>
+#include <cstdint>
 
 namespace llvm {
 class StringRef;
diff -ruN llvm-11.0.0.src/lib/Support/Triple.cpp llvm-11.0.0.src.aros/lib/Support/Triple.cpp
--- llvm-11.0.0.src/lib/Support/Triple.cpp	2020-10-07 10:10:48.000000000 +0000
+++ llvm-11.0.0.src.aros/lib/Support/Triple.cpp	2020-10-07 10:10:48.000000000 +0000
@@ -188,6 +188,7 @@
   case AMDPAL: return "amdpal";
   case Ananas: return "ananas";
   case CNK: return "cnk";
+  case AROS: return "aros";
   case CUDA: return "cuda";
   case CloudABI: return "cloudabi";
   case Contiki: return "contiki";
@@ -489,6 +490,7 @@
 static Triple::OSType parseOS(StringRef OSName) {
   return StringSwitch<Triple::OSType>(OSName)
     .StartsWith("ananas", Triple::Ananas)
+    .StartsWith("aros", Triple::AROS)
     .StartsWith("cloudabi", Triple::CloudABI)
     .StartsWith("darwin", Triple::Darwin)
     .StartsWith("dragonfly", Triple::DragonFly)
