diff -ruN grub-2.02/grub-core/fs/affs.c grub-2.02.aros/grub-core/fs/affs.c
--- grub-2.02/grub-core/fs/affs.c	2015-05-21 16:50:29.000000000 +0100
+++ grub-2.02.aros/grub-core/fs/affs.c	2017-04-24 12:30:25.000000000 +0100
@@ -202,7 +202,9 @@
 	goto fail;
 
       /* Make sure this is an affs filesystem.  */
-      if (grub_strncmp ((char *) (data->bblock.type), "DOS", 3) != 0
+      if (((grub_strncmp ((char *) (data->bblock.type), "DOS", 3) != 0)
+	  /* Test if the filesystem is a muFS filesystem.  */
+	  && (grub_strncmp ((char *) (data->bblock.type), "muF", 3) != 0))
 	  /* Test if the filesystem is a OFS filesystem.  */
 	  || !(data->bblock.flags & GRUB_AFFS_FLAG_FFS))
 	continue;
diff -ruN grub-2.02/grub-core/fs/btrfs.c grub-2.02.aros/grub-core/fs/btrfs.c
--- grub-2.02/grub-core/fs/btrfs.c	2017-04-24 10:16:00.000000000 +0100
+++ grub-2.02.aros/grub-core/fs/btrfs.c	2017-04-24 12:30:25.000000000 +0100
@@ -175,7 +175,7 @@
 {
   grub_int64_t sec;
   grub_uint32_t nanosec;
-} __attribute__ ((aligned (4)));
+} GRUB_PACKED;
 
 struct grub_btrfs_inode
 {
diff -ruN grub-2.02/grub-core/osdep/aros/getroot.c grub-2.02.aros/grub-core/osdep/aros/getroot.c
--- grub-2.02/grub-core/osdep/aros/getroot.c	2015-05-21 16:50:29.000000000 +0100
+++ grub-2.02.aros/grub-core/osdep/aros/getroot.c	2017-04-24 12:30:25.000000000 +0100
@@ -29,7 +29,6 @@
 
 #include <string.h>
 #include <dos/dos.h>
-#include <dos/filesystem.h>
 #include <dos/exall.h>
 #include <proto/dos.h>
 #include <proto/exec.h>
diff -ruN grub-2.02/grub-core/osdep/aros/hostdisk.c grub-2.02.aros/grub-core/osdep/aros/hostdisk.c
--- grub-2.02/grub-core/osdep/aros/hostdisk.c	2016-02-28 14:22:21.000000000 +0000
+++ grub-2.02.aros/grub-core/osdep/aros/hostdisk.c	2017-04-24 12:30:25.000000000 +0100
@@ -45,7 +45,6 @@
 
 #include <string.h>
 #include <dos/dos.h>
-#include <dos/filesystem.h>
 #include <dos/exall.h>
 #include <proto/dos.h>
 #include <devices/hardblocks.h>
diff -ruN grub-2.02/grub-core/osdep/aros/relpath.c grub-2.02.aros/grub-core/osdep/aros/relpath.c
--- grub-2.02/grub-core/osdep/aros/relpath.c	2015-05-21 16:50:29.000000000 +0100
+++ grub-2.02.aros/grub-core/osdep/aros/relpath.c	2017-04-24 12:30:25.000000000 +0100
@@ -29,7 +29,6 @@
 
 #include <string.h>
 #include <dos/dos.h>
-#include <dos/filesystem.h>
 #include <dos/exall.h>
 #include <proto/dos.h>
 #include <proto/exec.h>
diff -ruN grub-2.02/include/grub/efiemu/runtime.h grub-2.02.aros/include/grub/efiemu/runtime.h
--- grub-2.02/include/grub/efiemu/runtime.h	2015-05-21 16:50:29.000000000 +0100
+++ grub-2.02.aros/include/grub/efiemu/runtime.h	2019-03-29 16:16:01.449086200 +0000
@@ -29,7 +29,7 @@
 
 struct efi_variable
 {
-  grub_efi_guid_t guid;
+  grub_efi_packed_guid_t guid;
   grub_uint32_t namelen;
   grub_uint32_t size;
   grub_efi_uint32_t attributes;
diff -ruN grub-2.02/include/grub/gpt_partition.h grub-2.02.aros/include/grub/gpt_partition.h
--- grub-2.02/include/grub/gpt_partition.h	2015-05-21 16:50:29.000000000 +0100
+++ grub-2.02.aros/include/grub/gpt_partition.h	2019-03-29 16:16:48.443396000 +0000
@@ -28,7 +28,7 @@
   grub_uint16_t data2;
   grub_uint16_t data3;
   grub_uint8_t data4[8];
-} __attribute__ ((aligned(8)));
+} GRUB_PACKED;
 typedef struct grub_gpt_part_type grub_gpt_part_type_t;
 
 #define GRUB_GPT_PARTITION_TYPE_EMPTY \
diff -ruN grub-2.02/util/grub-mkimagexx.c grub-2.02.aros/util/grub-mkimagexx.c
--- grub-2.02/util/grub-mkimagexx.c	2017-04-24 10:16:00.000000000 +0100
+++ grub-2.02.aros/util/grub-mkimagexx.c	2017-04-24 12:30:25.000000000 +0100
@@ -832,6 +832,7 @@
 		  break;
 
 		case R_X86_64_PC32:
+		case R_X86_64_PLT32:
 		  {
 		    grub_uint32_t *t32 = (grub_uint32_t *) target;
 		    *t32 = grub_host_to_target64 (grub_target_to_host32 (*t32)
diff -ruN grub-2.02/util/grub-module-verifier.c grub-2.02.aros/util/grub-module-verifier.c
--- grub-2.02/util/grub-module-verifier.c	2017-04-24 10:16:00.000000000 +0100
+++ grub-2.02.aros/util/grub-module-verifier.c	2017-04-24 12:30:25.000000000 +0100
@@ -19,6 +19,7 @@
       -1
     }, (int[]){
       R_X86_64_PC32,
+      R_X86_64_PLT32,
       -1
     }
   },
