diff -ruN glu-9.0.2/src/include/gluos.h glu-9.0.2.aros/src/include/gluos.h
--- glu-9.0.2/src/include/gluos.h	2021-06-24 21:36:59.000000000 +0000
+++ glu-9.0.2.aros/src/include/gluos.h	2022-05-07 12:14:17.800645600 +0000
@@ -77,6 +77,8 @@
 #include <malloc.h>
 #define WINGDIAPI
 
+#elif defined(__AROS__)
+#include <stddef.h>
 #else
 
 /* Disable Microsoft-specific keywords */
diff -ruN glu-9.0.2/src/libnurbs/interface/insurfeval.cc glu-9.0.2.aros/src/libnurbs/interface/insurfeval.cc
--- glu-9.0.2/src/libnurbs/interface/insurfeval.cc	2021-06-24 21:36:59.000000000 +0000
+++ glu-9.0.2.aros/src/libnurbs/interface/insurfeval.cc	2022-05-07 12:14:17.800645600 +0000
@@ -1344,6 +1344,8 @@
   free(lowerXYZ);
   free(upperNormal);
   free(lowerNormal);
+
+  (void)leftMostV; // Unused
 }
 
 /*triangulate a strip bounded by two lines which are parallel  to V-axis
@@ -1534,6 +1536,8 @@
   free(rightXYZ);
   free(leftNormal);
   free(rightNormal);
+
+  (void)botMostV; // Unused
 }
 
 /*-----------------------begin evalMachine-------------------*/
diff -ruN glu-9.0.2/src/libnurbs/nurbtess/sampleCompTop.cc glu-9.0.2.aros/src/libnurbs/nurbtess/sampleCompTop.cc
--- glu-9.0.2/src/libnurbs/nurbtess/sampleCompTop.cc	2021-06-24 21:36:59.000000000 +0000
+++ glu-9.0.2.aros/src/libnurbs/nurbtess/sampleCompTop.cc	2022-05-07 12:14:17.800645600 +0000
@@ -511,26 +511,6 @@
     }
   else if(up_rightCornerWhere != 2)
     {
-      Real* tempTop;
-      Int tempLeftStart;
-      if(up_rightCornerWhere == 1)
-	{
-	  tempLeftStart = leftStartIndex;
-	  tempTop = topVertex;
-	}
-      else //0
-	{
-	  tempLeftStart = up_rightCornerIndex+1;
-	  tempTop = leftChain->getVertex(up_rightCornerIndex);
-	}
-/*
-      sampleTopLeftWithGridLine(tempTop, leftChain, tempLeftStart, up_leftCornerIndex,
-				leftGridChain->getGrid(),
-				 leftGridChain->getVlineIndex(gridIndex1),
-				 leftGridChain->getUlineIndex(gridIndex1),
-				 rightGridChain->getUlineIndex(gridIndex1),
-				 pStream);
-*/
       sampleCompTopSimple(topVertex,
 			  leftChain,
 			  leftStartIndex,
