diff -ruN glu-9.0.0/src/libnurbs/interface/insurfeval.cc glu-9.0.0.aros/src/libnurbs/interface/insurfeval.cc
--- glu-9.0.0/src/libnurbs/interface/insurfeval.cc	2012-09-14 20:48:54.000000000 -0400
+++ glu-9.0.0.aros/src/libnurbs/interface/insurfeval.cc	2015-04-19 17:13:37.212360145 -0400
@@ -1344,6 +1344,8 @@
   free(lowerXYZ);
   free(upperNormal);
   free(lowerNormal);
+
+  (void)leftMostV; // Unused
 }
 
 /*triangulate a strip bounded by two lines which are parallel  to V-axis
@@ -1534,6 +1536,8 @@
   free(rightXYZ);
   free(leftNormal);
   free(rightNormal);
+
+  (void)botMostV; // Unused
 }
 
 /*-----------------------begin evalMachine-------------------*/
diff -ruN glu-9.0.0/src/libnurbs/nurbtess/sampleCompTop.cc glu-9.0.0.aros/src/libnurbs/nurbtess/sampleCompTop.cc
--- glu-9.0.0/src/libnurbs/nurbtess/sampleCompTop.cc	2012-09-14 20:48:54.000000000 -0400
+++ glu-9.0.0.aros/src/libnurbs/nurbtess/sampleCompTop.cc	2015-04-19 17:14:42.942805024 -0400
@@ -511,26 +511,6 @@
     }
   else if(up_rightCornerWhere != 2)
     {
-      Real* tempTop;
-      Int tempLeftStart;
-      if(up_rightCornerWhere == 1)
-	{
-	  tempLeftStart = leftStartIndex;
-	  tempTop = topVertex;
-	}
-      else //0
-	{
-	  tempLeftStart = up_rightCornerIndex+1;
-	  tempTop = leftChain->getVertex(up_rightCornerIndex);
-	}
-/*
-      sampleTopLeftWithGridLine(tempTop, leftChain, tempLeftStart, up_leftCornerIndex,
-				leftGridChain->getGrid(),
-				 leftGridChain->getVlineIndex(gridIndex1),
-				 leftGridChain->getUlineIndex(gridIndex1),
-				 rightGridChain->getUlineIndex(gridIndex1),
-				 pStream);
-*/
       sampleCompTopSimple(topVertex,
 			  leftChain,
 			  leftStartIndex,
