diff -ruN glu-9.0.0/src/libnurbs/interface/insurfeval.cc glu-9.0.0.aros/src/libnurbs/interface/insurfeval.cc
--- glu-9.0.0/src/libnurbs/interface/insurfeval.cc	2012-09-15 01:48:54.000000000 +0100
+++ glu-9.0.0.aros/src/libnurbs/interface/insurfeval.cc	2019-04-03 20:07:20.650123200 +0100
@@ -1344,6 +1344,8 @@
   free(lowerXYZ);
   free(upperNormal);
   free(lowerNormal);
+
+  (void)leftMostV; // Unused
 }
 
 /*triangulate a strip bounded by two lines which are parallel  to V-axis
@@ -1534,6 +1536,8 @@
   free(rightXYZ);
   free(leftNormal);
   free(rightNormal);
+
+  (void)botMostV; // Unused
 }
 
 /*-----------------------begin evalMachine-------------------*/
diff -ruN glu-9.0.0/src/libnurbs/nurbtess/sampleCompTop.cc glu-9.0.0.aros/src/libnurbs/nurbtess/sampleCompTop.cc
--- glu-9.0.0/src/libnurbs/nurbtess/sampleCompTop.cc	2012-09-15 01:48:54.000000000 +0100
+++ glu-9.0.0.aros/src/libnurbs/nurbtess/sampleCompTop.cc	2019-04-03 20:07:20.655097500 +0100
@@ -511,26 +511,6 @@
     }
   else if(up_rightCornerWhere != 2)
     {
-      Real* tempTop;
-      Int tempLeftStart;
-      if(up_rightCornerWhere == 1)
-	{
-	  tempLeftStart = leftStartIndex;
-	  tempTop = topVertex;
-	}
-      else //0
-	{
-	  tempLeftStart = up_rightCornerIndex+1;
-	  tempTop = leftChain->getVertex(up_rightCornerIndex);
-	}
-/*
-      sampleTopLeftWithGridLine(tempTop, leftChain, tempLeftStart, up_leftCornerIndex,
-				leftGridChain->getGrid(),
-				 leftGridChain->getVlineIndex(gridIndex1),
-				 leftGridChain->getUlineIndex(gridIndex1),
-				 rightGridChain->getUlineIndex(gridIndex1),
-				 pStream);
-*/
       sampleCompTopSimple(topVertex,
 			  leftChain,
 			  leftStartIndex,
diff -ruN glu-9.0.0/src/libtess/priorityq.c glu-9.0.0.aros/src/libtess/priorityq.c
--- glu-9.0.0/src/libtess/priorityq.c	2012-09-15 01:48:54.000000000 +0100
+++ glu-9.0.0.aros/src/libtess/priorityq.c	2019-04-10 03:14:26.089025000 +0100
@@ -65,6 +65,7 @@
      return NULL;
   }
 
+  pq->order = NULL;
   pq->size = 0;
   pq->max = INIT_SIZE;
   pq->initialized = FALSE;
diff -ruN glu-9.0.0/src/libutil/mipmap.c glu-9.0.0.aros/src/libutil/mipmap.c
--- glu-9.0.0/src/libutil/mipmap.c	2012-09-15 01:48:54.000000000 +0100
+++ glu-9.0.0.aros/src/libutil/mipmap.c	2019-04-10 03:13:14.787011600 +0100
@@ -4509,6 +4509,8 @@
 	     glPixelStorei(GL_UNPACK_SKIP_PIXELS, psm.unpack_skip_pixels);
 	     glPixelStorei(GL_UNPACK_ROW_LENGTH, psm.unpack_row_length);
 	     glPixelStorei(GL_UNPACK_SWAP_BYTES, psm.unpack_swap_bytes);
+	     free(srcImage);
+	     free(dstImage);
 	     return GLU_OUT_OF_MEMORY;
 	  }
 
