diff -ur drm-2.4.26/nouveau/nouveau_bo.c libdrm/nouveau/nouveau_bo.c
--- drm-2.4.26/nouveau/nouveau_bo.c	2011-06-04 05:01:50.000000000 +0200
+++ libdrm/nouveau/nouveau_bo.c	2011-09-01 17:11:21.000000000 +0200
@@ -28,7 +28,9 @@
 #include <errno.h>
 #include <assert.h>
 
+#if !defined(__AROS__)
 #include <sys/mman.h>
+#endif
 
 #include "nouveau_private.h"
 
@@ -100,7 +102,11 @@
 		return;
 
 	if (nvbo->map) {
+#if !defined(__AROS__)        
 		munmap(nvbo->map, nvbo->size);
+#else
+        drmMUnmap(nvdev->fd, nvbo->handle);
+#endif
 		nvbo->map = NULL;
 	}
 
@@ -166,13 +172,18 @@
 	if (!nvbo->map_handle)
 		return -EINVAL;
 
+#if !defined(__AROS__)
 	nvbo->map = mmap(0, nvbo->size, PROT_READ | PROT_WRITE,
 			 MAP_SHARED, nvdev->fd, nvbo->map_handle);
 	if (nvbo->map == MAP_FAILED) {
 		nvbo->map = NULL;
 		return -errno;
 	}
-
+#else
+    nvbo->map = drmMMap(nvdev->fd, nvbo->handle);
+    if (nvbo->map == NULL)
+        return -EINVAL;
+#endif
 	return 0;
 }
 
@@ -383,8 +394,6 @@
 		req.flags |= NOUVEAU_GEM_CPU_PREP_WRITE;
 	if (no_wait)
 		req.flags |= NOUVEAU_GEM_CPU_PREP_NOWAIT;
-	if (no_block)
-		req.flags |= NOUVEAU_GEM_CPU_PREP_NOBLOCK;
 
 	do {
 		ret = drmCommandWrite(nvdev->fd, DRM_NOUVEAU_GEM_CPU_PREP,
diff -ur drm-2.4.26/nouveau/nouveau_drmif.h libdrm/nouveau/nouveau_drmif.h
--- drm-2.4.26/nouveau/nouveau_drmif.h	2011-06-04 05:01:50.000000000 +0200
+++ libdrm/nouveau/nouveau_drmif.h	2011-09-01 17:11:21.000000000 +0200
@@ -24,7 +24,7 @@
 #define __NOUVEAU_DRMIF_H__
 
 #include <stdint.h>
-#include <xf86drm.h>
+#include <arosdrm.h>
 
 #include "nouveau_device.h"
 
diff -ur drm-2.4.26/nouveau/nouveau_private.h libdrm/nouveau/nouveau_private.h
--- drm-2.4.26/nouveau/nouveau_private.h	2011-06-04 05:01:50.000000000 +0200
+++ libdrm/nouveau/nouveau_private.h	2011-09-01 17:11:21.000000000 +0200
@@ -24,7 +24,7 @@
 #define __NOUVEAU_PRIVATE_H__
 
 #include <stdint.h>
-#include <xf86drm.h>
+#include <arosdrm.h>
 #include <nouveau_drm.h>
 
 #include "nouveau_drmif.h"
