diff -urN binutils-2.21.1.orig/bfd/config.bfd binutils-2.21.1/bfd/config.bfd
--- binutils-2.21.1.orig/bfd/config.bfd	2011-04-01 08:02:42.000000000 -0400
+++ binutils-2.21.1/bfd/config.bfd	2011-11-18 12:12:30.000000000 -0500
@@ -281,7 +281,7 @@
     ;;
   arm-*-elf | arm-*-freebsd* | arm*-*-linux-* | arm*-*-conix* | \
   arm*-*-uclinux* | arm-*-kfreebsd*-gnu | \
-  arm*-*-eabi* )
+  arm*-*-eabi* | arm-*-aros)
     targ_defvec=bfd_elf32_littlearm_vec
     targ_selvecs=bfd_elf32_bigarm_vec
     ;;
@@ -518,6 +518,8 @@
     ;;
   i[3-7]86-*-aros*)
     targ_defvec=bfd_elf32_i386_vec
+    targ_selvecs="bfd_elf64_x86_64_vec bfd_elf32_powerpc_vec"
+    want64=true
     ;;
   i[3-7]86-*-chorus*)
     targ_defvec=bfd_elf32_i386_vec
@@ -647,6 +649,11 @@
     want64=true
     targ_underscore=no
     ;;
+  x86_64-*-aros*)
+    targ_defvec=bfd_elf64_x86_64_vec
+    targ_selvecs="bfd_elf32_i386_vec bfd_elf32_powerpc_vec"
+    want64=true
+    ;;
 #endif
   i[3-7]86-*-lynxos*)
     targ_defvec=bfd_elf32_i386_vec
@@ -802,7 +809,7 @@
     targ_selvecs="cisco_core_big_vec ieee_vec"
     targ_underscore=yes
     ;;
-  m68*-*-elf* | m68*-*-sysv4* | m68*-*-uclinux*)
+  m68*-*-elf* | m68*-*-sysv4* | m68*-*-uclinux* | m68*-*-aros*)
     targ_defvec=bfd_elf32_m68k_vec
     targ_selvecs="m68kcoff_vec ieee_vec"
     ;;
@@ -1197,6 +1204,11 @@
     targ_defvec=bfd_elf32_rx_le_vec
     targ_selvecs="bfd_elf32_rx_be_vec bfd_elf32_rx_le_vec"
     ;;
+  powerpc-*-aros*)
+    targ_defvec=bfd_elf32_powerpc_vec
+    targ_selvecs="bfd_elf32_i386_vec bfd_elf64_x86_64_vec"
+    want64=true
+    ;;
 
   s390-*-linux*)
     targ_defvec=bfd_elf32_s390_vec
diff -urN binutils-2.21.1.orig/config.guess binutils-2.21.1/config.guess
--- binutils-2.21.1.orig/config.guess	2010-06-01 13:53:40.000000000 -0400
+++ binutils-2.21.1/config.guess	2011-06-27 03:39:00.000000000 -0400
@@ -289,6 +289,9 @@
     *:[Mm]orph[Oo][Ss]:*:*)
 	echo ${UNAME_MACHINE}-unknown-morphos
 	exit ;;
+    *:AROS:*:*)
+	echo ${UNAME_MACHINE}-unknown-aros
+	exit ;;
     *:OS/390:*:*)
 	echo i370-ibm-openedition
 	exit ;;
diff -urN binutils-2.21.1.orig/intl/configure binutils-2.21.1/intl/configure
--- binutils-2.21.1.orig/intl/configure	2010-11-05 06:31:38.000000000 -0400
+++ binutils-2.21.1/intl/configure	2011-06-27 03:39:00.000000000 -0400
@@ -4606,6 +4606,9 @@
 
 int main ()
 {
+#if defined(__AROS__)
+  exit (1);
+#endif
   signal (SIGFPE, sigfpe_handler);
 /* IRIX and AIX (when "xlc -qcheck" is used) yield signal SIGTRAP.  */
 #if (defined (__sgi) || defined (_AIX)) && defined (SIGTRAP)
diff -urN binutils-2.21.1.orig/ld/configure.tgt binutils-2.21.1/ld/configure.tgt
--- binutils-2.21.1.orig/ld/configure.tgt	2011-04-01 08:03:54.000000000 -0400
+++ binutils-2.21.1/ld/configure.tgt	2011-11-18 12:13:44.000000000 -0500
@@ -79,6 +79,10 @@
 			targ_extra_emuls="armelfb armelf armelf_linux"
 			targ_extra_libpath="armelf_linux"
 			;;
+arm*-*-aros*)		targ_emul=armelf_linux_eabi
+			targ_extra_emuls=armelfb_linux_eabi
+			targ_extra_libpath=$targ_extra_emuls
+			;;
 arm*-*-linux-*eabi)	targ_emul=armelf_linux_eabi
 			targ_extra_emuls=armelfb_linux_eabi
 			targ_extra_libpath=$targ_extra_emuls
@@ -245,6 +249,9 @@
 x86_64-*-elf*)		targ_emul=elf_x86_64
 			targ_extra_emuls="elf_i386 elf_l1om"
 			;;
+x86_64-*-aros*)		targ_emul=elf_x86_64
+			targ_extra_emuls=elf_i386
+			;;
 i[3-7]86-*-kaos*)	targ_emul=elf_i386 ;;
 i[3-7]86-*-freebsdaout* | i[3-7]86-*-freebsd[12].* | i[3-7]86-*-freebsd[12])
 			targ_emul=i386bsd ;;
@@ -336,6 +343,7 @@
 m68k-hp-bsd*)		targ_emul=hp300bsd ;;
 m68*-motorola-sysv*)	targ_emul=delta68 ;;
 m68*-*-aout)		targ_emul=m68kaout ;;
+m68*-*-aros)		targ_emul=m68kelf ;;
 m68*-*-coff)		targ_emul=m68kcoff ;;
 m68*-*-elf)		targ_emul=m68kelf ;;
 m68*-*-hpux*)		targ_emul=hp3hpux ;;
@@ -504,6 +512,7 @@
 powerpc-*-beos*)	targ_emul=aixppc ;;
 powerpc-*-windiss*)	targ_emul=elf32ppcwindiss ;;
 powerpc-*-lynxos*)	targ_emul=ppclynx ;;
+powerpc-*-aros*)	targ_emul=elf32ppc ;;
 rs6000-*-aix[5-9]*)	targ_emul=aix5rs6 ;;
 rs6000-*-aix*)		targ_emul=aixrs6
 			;;
