diff -ru binutils-2.20.1.orig/bfd/config.bfd binutils-2.20.1/bfd/config.bfd
--- binutils-2.20.1.orig/bfd/config.bfd	2009-08-06 21:38:00.000000000 +0400
+++ binutils-2.20.1/bfd/config.bfd	2010-10-20 08:24:36.000000000 +0400
@@ -282,7 +282,7 @@
     ;;
   arm-*-elf | arm-*-freebsd* | arm*-*-linux-* | arm*-*-conix* | \
   arm*-*-uclinux* | arm-*-kfreebsd*-gnu | \
-  arm*-*-eabi* )
+  arm*-*-eabi* | arm-*-aros)
     targ_defvec=bfd_elf32_littlearm_vec
     targ_selvecs=bfd_elf32_bigarm_vec
     ;;
@@ -521,6 +521,8 @@
     ;;
   i[3-7]86-*-aros*)
     targ_defvec=bfd_elf32_i386_vec
+    targ_selvecs="bfd_elf64_x86_64_vec bfd_elf32_powerpc_vec"
+    want64=true
     ;;
   i[3-7]86-*-chorus*)
     targ_defvec=bfd_elf32_i386_vec
@@ -641,6 +643,11 @@
     want64=true
     targ_underscore=yes
     ;;
+  x86_64-*-aros*)
+    targ_defvec=bfd_elf64_x86_64_vec
+    targ_selvecs="bfd_elf32_i386_vec bfd_elf32_powerpc_vec"
+    want64=true
+    ;;
 #endif
   i[3-7]86-*-lynxos*)
     targ_defvec=bfd_elf32_i386_vec
@@ -1190,6 +1197,11 @@
     targ_defvec=bfd_powerpcle_pe_vec
     targ_selvecs="bfd_powerpcle_pei_vec bfd_powerpc_pei_vec bfd_powerpcle_pe_vec bfd_powerpc_pe_vec"
     ;;
+  powerpc-*-aros*)
+    targ_defvec=bfd_elf32_powerpc_vec
+    targ_selvecs="bfd_elf32_i386_vec bfd_elf64_x86_64_vec"
+    want64=true
+    ;;
 
   s390-*-linux*)
     targ_defvec=bfd_elf32_s390_vec
diff -ru binutils-2.20.1.orig/config.guess binutils-2.20.1/config.guess
--- binutils-2.20.1.orig/config.guess	2009-08-17 08:10:29.000000000 +0400
+++ binutils-2.20.1/config.guess	2010-10-20 08:20:36.000000000 +0400
@@ -288,6 +288,9 @@
     *:[Mm]orph[Oo][Ss]:*:*)
 	echo ${UNAME_MACHINE}-unknown-morphos
 	exit ;;
+    *:AROS:*:*)
+	echo ${UNAME_MACHINE}-unknown-aros
+	exit ;;
     *:OS/390:*:*)
 	echo i370-ibm-openedition
 	exit ;;
diff -ru binutils-2.20.1.orig/intl/configure binutils-2.20.1/intl/configure
--- binutils-2.20.1.orig/intl/configure	2010-03-03 16:52:20.000000000 +0300
+++ binutils-2.20.1/intl/configure	2010-10-20 08:20:36.000000000 +0400
@@ -4606,6 +4606,9 @@
 
 int main ()
 {
+#if defined(__AROS__)
+  exit (1);
+#endif
   signal (SIGFPE, sigfpe_handler);
 /* IRIX and AIX (when "xlc -qcheck" is used) yield signal SIGTRAP.  */
 #if (defined (__sgi) || defined (_AIX)) && defined (SIGTRAP)
diff -ru binutils-2.20.1.orig/ld/configure.tgt binutils-2.20.1/ld/configure.tgt
--- binutils-2.20.1.orig/ld/configure.tgt	2010-10-18 18:10:17.000000000 +0400
+++ binutils-2.20.1/ld/configure.tgt	2010-10-20 08:29:18.000000000 +0400
@@ -78,6 +78,10 @@
 			targ_extra_emuls="armelfb armelf armelf_linux"
 			targ_extra_libpath="armelf_linux"
 			;;
+arm*-*-aros*)		targ_emul=armelf_linux_eabi
+			targ_extra_emuls=armelfb_linux_eabi
+			targ_extra_libpath=$targ_extra_emuls
+			;;
 arm*-*-linux-*eabi)	targ_emul=armelf_linux_eabi
 			targ_extra_emuls=armelfb_linux_eabi
 			targ_extra_libpath=$targ_extra_emuls
@@ -244,6 +248,9 @@
 x86_64-*-elf*)		targ_emul=elf_x86_64
 			targ_extra_emuls="elf_i386 elf_l1om"
 			;;
+x86_64-*-aros*)		targ_emul=elf_x86_64
+			targ_extra_emuls=elf_i386
+			;;
 i[3-7]86-*-kaos*)	targ_emul=elf_i386 ;;
 i[3-7]86-*-freebsdaout* | i[3-7]86-*-freebsd[12].* | i[3-7]86-*-freebsd[12])
 			targ_emul=i386bsd ;;
@@ -500,6 +507,7 @@
 powerpc-*-beos*)	targ_emul=aixppc ;;
 powerpc-*-windiss*)	targ_emul=elf32ppcwindiss ;;
 powerpc-*-lynxos*)	targ_emul=ppclynx ;;
+powerpc-*-aros*)	targ_emul=elf32ppc ;;
 rs6000-*-aix[5-9]*)	targ_emul=aix5rs6 ;;
 rs6000-*-aix*)		targ_emul=aixrs6
 			;;
