diff -ruN binutils-2.23.2/bfd/config.bfd binutils-2.23.2.aros/bfd/config.bfd
--- binutils-2.23.2/bfd/config.bfd	2012-09-04 15:14:59.000000000 +0100
+++ binutils-2.23.2.aros/bfd/config.bfd	2012-06-28 12:50:52.000000000 +0100
@@ -315,7 +315,7 @@
     ;;
   arm-*-elf | arm-*-freebsd* | arm*-*-linux-* | arm*-*-conix* | \
   arm*-*-uclinux* | arm-*-kfreebsd*-gnu | \
-  arm*-*-eabi* )
+  arm*-*-eabi* | arm-*-aros)
     targ_defvec=bfd_elf32_littlearm_vec
     targ_selvecs=bfd_elf32_bigarm_vec
     ;;
@@ -518,6 +518,8 @@
     ;;
   i[3-7]86-*-aros*)
     targ_defvec=bfd_elf32_i386_vec
+    targ_selvecs="bfd_elf64_x86_64_vec bfd_elf32_powerpc_vec"
+    want64=true
     ;;
   i[3-7]86-*-chorus*)
     targ_defvec=bfd_elf32_i386_vec
@@ -660,6 +662,11 @@
     want64=true
     targ_underscore=no
     ;;
+  x86_64-*-aros*)
+    targ_defvec=bfd_elf64_x86_64_vec
+    targ_selvecs="bfd_elf32_i386_vec bfd_elf32_powerpc_vec"
+    want64=true
+    ;;
 #endif
   i[3-7]86-*-lynxos*)
     targ_defvec=bfd_elf32_i386_vec
@@ -815,7 +822,7 @@
     targ_selvecs="cisco_core_big_vec ieee_vec"
     targ_underscore=yes
     ;;
-  m68*-*-elf* | m68*-*-sysv4* | m68*-*-uclinux*)
+  m68*-*-elf* | m68*-*-sysv4* | m68*-*-uclinux* | m68*-*-aros)
     targ_defvec=bfd_elf32_m68k_vec
     targ_selvecs="m68kcoff_vec ieee_vec"
     ;;
@@ -1190,6 +1197,11 @@
     want64=true
     ;;
 #endif
+  powerpc-*-aros*)
+    targ_defvec=bfd_elf32_powerpc_vec
+    targ_selvecs="bfd_elf32_i386_vec bfd_elf64_x86_64_vec"
+    want64=true
+    ;;
   powerpc-*-*freebsd*)
     targ_defvec=bfd_elf32_powerpc_freebsd_vec
     targ_selvecs="rs6000coff_vec bfd_elf32_powerpc_vec bfd_elf32_powerpcle_vec ppcboot_vec"
diff -ruN binutils-2.23.2/binutils/configure binutils-2.23.2.aros/binutils/configure
--- binutils-2.23.2/binutils/configure	2012-11-05 16:27:34.000000000 +0000
+++ binutils-2.23.2.aros/binutils/configure	2013-07-17 11:44:59.680499534 +0100
@@ -12746,7 +12746,9 @@
 $as_echo_n "checking for mbstate_t... " >&6; }
   cat confdefs.h - <<_ACEOF >conftest.$ac_ext
 /* end confdefs.h.  */
+#if !defined(__AROS__)
 #include <wchar.h>
+#endif
 int
 main ()
 {
diff -ruN binutils-2.23.2/binutils/sysdep.h binutils-2.23.2.aros/binutils/sysdep.h
--- binutils-2.23.2/binutils/sysdep.h	2012-05-11 15:25:29.000000000 +0100
+++ binutils-2.23.2.aros/binutils/sysdep.h	2013-07-17 11:09:53.252362151 +0100
@@ -26,6 +26,7 @@
 #include "ansidecl.h"
 #include <stdio.h>
 #include <sys/types.h>
+#include <sys/stat.h>
 
 #include "bfdver.h"
 
diff -ruN binutils-2.23.2/config.guess binutils-2.23.2.aros/config.guess
--- binutils-2.23.2/config.guess	2011-06-06 11:36:06.000000000 +0100
+++ binutils-2.23.2.aros/config.guess	2012-06-28 12:50:52.000000000 +0100
@@ -292,6 +292,9 @@
     *:[Mm]orph[Oo][Ss]:*:*)
 	echo ${UNAME_MACHINE}-unknown-morphos
 	exit ;;
+    *:AROS:*:*)
+	echo ${UNAME_MACHINE}-unknown-aros
+	exit ;;
     *:OS/390:*:*)
 	echo i370-ibm-openedition
 	exit ;;
diff -ruN binutils-2.23.2/intl/configure binutils-2.23.2.aros/intl/configure
--- binutils-2.23.2/intl/configure	2010-09-27 21:22:53.000000000 +0100
+++ binutils-2.23.2.aros/intl/configure	2012-06-28 12:50:52.000000000 +0100
@@ -4606,6 +4606,9 @@
 
 int main ()
 {
+#if defined(__AROS__)
+  exit (1);
+#endif
   signal (SIGFPE, sigfpe_handler);
 /* IRIX and AIX (when "xlc -qcheck" is used) yield signal SIGTRAP.  */
 #if (defined (__sgi) || defined (_AIX)) && defined (SIGTRAP)
diff -ruN binutils-2.23.2/ld/configure.tgt binutils-2.23.2.aros/ld/configure.tgt
--- binutils-2.23.2/ld/configure.tgt	2012-09-04 13:53:47.000000000 +0100
+++ binutils-2.23.2.aros/ld/configure.tgt	2012-06-28 12:50:52.000000000 +0100
@@ -87,6 +87,10 @@
 			targ_extra_emuls="armelfb armelf armelf_linux"
 			targ_extra_libpath="armelf_linux"
 			;;
+arm*-*-aros*)		targ_emul=armelf_linux_eabi
+			targ_extra_emuls=armelfb_linux_eabi
+			targ_extra_libpath=$targ_extra_emuls
+			;;
 arm*-*-linux-*eabi*)	targ_emul=armelf_linux_eabi
 			targ_extra_emuls=armelfb_linux_eabi
 			targ_extra_libpath=$targ_extra_emuls
@@ -260,6 +264,9 @@
 x86_64-*-elf*)		targ_emul=elf_x86_64
 			targ_extra_emuls="elf_i386 elf_l1om elf_k1om"
 			;;
+x86_64-*-aros*)		targ_emul=elf_x86_64
+			targ_extra_emuls=elf_i386
+			;;
 i[3-7]86-*-kaos*)	targ_emul=elf_i386 ;;
 i[3-7]86-*-freebsdaout* | i[3-7]86-*-freebsd[12].* | i[3-7]86-*-freebsd[12])
 			targ_emul=i386bsd ;;
@@ -366,6 +373,7 @@
 m68k-hp-bsd*)		targ_emul=hp300bsd ;;
 m68*-motorola-sysv*)	targ_emul=delta68 ;;
 m68*-*-aout)		targ_emul=m68kaout ;;
+m68*-*-aros)		targ_emul=m68kelf ;;
 m68*-*-coff)		targ_emul=m68kcoff ;;
 m68*-*-elf)		targ_emul=m68kelf ;;
 m68*-*-hpux*)		targ_emul=hp3hpux ;;
@@ -557,6 +565,7 @@
 powerpc-*-beos*)	targ_emul=aixppc ;;
 powerpc-*-windiss*)	targ_emul=elf32ppcwindiss ;;
 powerpc-*-lynxos*)	targ_emul=ppclynx ;;
+powerpc-*-aros*)	targ_emul=elf32ppc ;;
 rs6000-*-aix[5-9]*)	targ_emul=aix5rs6 ;;
 rs6000-*-aix*)		targ_emul=aixrs6
 			;;
