diff -ruN binutils-2.25/bfd/config.bfd binutils-2.25.aros/bfd/config.bfd
--- binutils-2.25/bfd/config.bfd	2012-09-04 16:14:59.000000000 +0200
+++ binutils-2.25.aros/bfd/config.bfd	2012-06-28 13:50:52.000000000 +0200
@@ -338,7 +338,7 @@
     ;;
   arm-*-elf | arm-*-freebsd* | arm*-*-linux-* | arm*-*-conix* | \
   arm*-*-uclinux* | arm-*-kfreebsd*-gnu | \
-  arm*-*-eabi* )
+  arm*-*-eabi* | arm-*-aros)
     targ_defvec=arm_elf32_le_vec
     targ_selvecs=arm_elf32_be_vec
     ;;
@@ -546,6 +546,8 @@
     ;;
   i[3-7]86-*-aros*)
     targ_defvec=i386_elf32_vec
+    targ_selvecs="x86_64_elf64_vec powerpc_elf32_vec"
+    want64=true
     ;;
   i[3-7]86-*-chorus*)
     targ_defvec=i386_elf32_vec
@@ -692,6 +694,11 @@
    targ_defvec=x86_64_elf64_vec
     want64=true
     ;;
+  x86_64-*-aros*)
+    targ_defvec=x86_64_elf64_vec
+    targ_selvecs="i386_elf32_vec powerpc_elf32_vec"
+    want64=true
+    ;;
 #endif
   i[3-7]86-*-lynxos*)
     targ_defvec=bfd_elf32_i386_vec
@@ -854,7 +854,7 @@
     targ_selvecs="core_cisco_be_vec ieee_vec"
     targ_underscore=yes
     ;;
-  m68*-*-elf* | m68*-*-sysv4* | m68*-*-uclinux*)
+  m68*-*-elf* | m68*-*-sysv4* | m68*-*-uclinux* | m68*-*-aros)
     targ_defvec=m68k_elf32_vec
     targ_selvecs="m68k_coff_vec ieee_vec"
     ;;
@@ -1264,6 +1264,11 @@
     want64=true
     ;;
 #endif
+  powerpc-*-aros*)
+    targ_defvec=powerpc_elf32_vec
+    targ_selvecs="i386_elf32_vec x86_64_elf64_vec"
+    want64=true
+    ;;
   powerpc-*-*freebsd*)
     targ_defvec=powerpc_elf32_freebsd_vec
     targ_selvecs="rs6000_xcoff_vec powerpc_elf32_vec powerpc_elf32_le_vec powerpc_boot_vec"
diff -ruN binutils-2.25/binutils/configure binutils-2.25.aros/binutils/configure
--- binutils-2.25/binutils/configure	2012-11-05 17:27:34.000000000 +0100
+++ binutils-2.25.aros/binutils/configure	2012-06-28 13:50:52.000000000 +0200
@@ -12875,7 +12875,9 @@
 $as_echo_n "checking for mbstate_t... " >&6; }
   cat confdefs.h - <<_ACEOF >conftest.$ac_ext
 /* end confdefs.h.  */
+#if !defined(__AROS__)
 #include <wchar.h>
+#endif
 int
 main ()
 {
diff -ruN binutils-2.25/binutils/sysdep.h binutils-2.25.aros/binutils/sysdep.h
--- binutils-2.25/binutils/sysdep.h	2012-05-11 16:25:29.000000000 +0200
+++ binutils-2.25.aros/binutils/sysdep.h	2012-06-28 13:50:52.000000000 +0200
@@ -24,6 +24,7 @@
 #include "ansidecl.h"
 #include <stdio.h>
 #include <sys/types.h>
+#include <sys/stat.h>
 
 #include "bfdver.h"
 
diff -ruN binutils-2.25/config.guess binutils-2.25.aros/config.guess
--- binutils-2.25/config.guess	2011-06-06 12:36:06.000000000 +0200
+++ binutils-2.25.aros/config.guess	2012-06-28 13:50:52.000000000 +0200
@@ -309,6 +309,9 @@
     *:[Mm]orph[Oo][Ss]:*:*)
 	echo ${UNAME_MACHINE}-unknown-morphos
 	exit ;;
+    *:AROS:*:*)
+	echo ${UNAME_MACHINE}-unknown-aros
+	exit ;;
     *:OS/390:*:*)
 	echo i370-ibm-openedition
 	exit ;;
diff -ruN binutils-2.25/intl/configure binutils-2.25.aros/intl/configure
--- binutils-2.25/intl/configure	2010-09-27 22:22:53.000000000 +0200
+++ binutils-2.25.aros/intl/configure	2012-06-28 13:50:52.000000000 +0200
@@ -4606,6 +4606,9 @@
 
 int main ()
 {
+#if defined(__AROS__)
+  exit (1);
+#endif
   signal (SIGFPE, sigfpe_handler);
 /* IRIX and AIX (when "xlc -qcheck" is used) yield signal SIGTRAP.  */
 #if (defined (__sgi) || defined (_AIX)) && defined (SIGTRAP)
diff -ruN binutils-2.25/ld/configure.tgt binutils-2.25.aros/ld/configure.tgt
--- binutils-2.25/ld/configure.tgt	2012-09-04 14:53:47.000000000 +0200
+++ binutils-2.25.aros/ld/configure.tgt	2012-06-28 13:50:52.000000000 +0200
@@ -110,6 +110,10 @@
 			targ_extra_emuls="armelfb armelf armelf_linux"
 			targ_extra_libpath="armelf_linux"
 			;;
+arm*-*-aros*)		targ_emul=armelf_aros
+			targ_extra_emuls=armelfb_aros
+			targ_extra_libpath=$targ_extra_emuls
+			;;
 arm*-*-linux-*eabi*)	targ_emul=armelf_linux_eabi
 			targ_extra_emuls=armelfb_linux_eabi
 			targ_extra_libpath=$targ_extra_emuls
@@ -293,6 +293,9 @@
 			targ_extra_libpath="elf_i386 elf32_x86_64 elf_l1om elf_k1om"
 			tdir_elf_i386=`echo ${targ_alias} | sed -e 's/x86_64/i386/'`
 			;;
+x86_64-*-aros*)		targ_emul=elf_x86_64
+			targ_extra_emuls=elf_i386
+			;;
 i[3-7]86-*-kaos*)	targ_emul=elf_i386 ;;
 i[3-7]86-*-freebsdaout* | i[3-7]86-*-freebsd[12].* | i[3-7]86-*-freebsd[12])
 			targ_emul=i386bsd ;;
@@ -406,6 +406,7 @@
 m68k-hp-bsd*)		targ_emul=hp300bsd ;;
 m68*-motorola-sysv*)	targ_emul=delta68 ;;
 m68*-*-aout)		targ_emul=m68kaout ;;
+m68*-*-aros)		targ_emul=m68kelf ;;
 m68*-*-coff)		targ_emul=m68kcoff ;;
 m68*-*-elf)		targ_emul=m68kelf ;;
 m68*-*-hpux*)		targ_emul=hp3hpux ;;
@@ -612,6 +612,7 @@
 powerpc-*-beos*)	targ_emul=aixppc ;;
 powerpc-*-windiss*)	targ_emul=elf32ppcwindiss ;;
 powerpc-*-lynxos*)	targ_emul=ppclynx ;;
+powerpc-*-aros*)	targ_emul=elf32ppc ;;
 rs6000-*-aix[5-9]*)	targ_emul=aix5rs6 ;;
 rs6000-*-aix*)		targ_emul=aixrs6
 			;;
diff -ruN binutils-2.25/ld/emulparams/armelf_aros.sh binutils-2.25.aros/ld/emulparams/armelf_aros.sh
--- binutils-2.25/ld/emulparams/armelf_aros.sh	1970-01-01 01:00:00.000000000 +0100
+++ binutils-2.25.aros/ld/emulparams/armelf_aros.sh	2012-06-28 13:50:52.000000000 +0200
@@ -0,0 +1,3 @@
+. ${srcdir}/emulparams/armelf_linux_eabi.sh
+# Remap R_ARM_TARGET2 to R_ARM_ABS32
+TARGET2_TYPE=abs
diff -ruN binutils-2.25/ld/emulparams/armelfb_aros.sh binutils-2.25.aros/ld/emulparams/armelfb_aros.sh
--- binutils-2.25/ld/emulparams/armelfb_aros.sh	1970-01-01 01:00:00.000000000 +0100
+++ binutils-2.25.aros/ld/emulparams/armelfb_aros.sh	2012-06-28 13:50:52.000000000 +0200
@@ -0,0 +1,2 @@
+. ${srcdir}/emulparams/armelf_aros.sh
+OUTPUT_FORMAT="elf32-bigarm"
diff -ruN binutils-2.25/ld/Makefile.in binutils-2.25.aros/ld/Makefile.in
--- binutils-2.25/ld/Makefile.in	2013-03-25 09:06:23.000000000 +0100
+++ binutils-2.25 .aros/ld/Makefile.in	2012-06-28 13:50:52.000000000 +0200
@@ -468,6 +468,7 @@
 	earmaoutl.c \
 	earmcoff.c \
 	earmelf.c \
+	earmelf_aros.c \
 	earmelf_fbsd.c \
 	earmelf_linux.c \
 	earmelf_linux_eabi.c \
@@ -476,6 +476,7 @@
 	earmelf_nbsd.c \
 	earmelf_vxworks.c \
 	earmelfb.c \
+	earmelfb_aros.c \
 	earmelfb_fbsd.c \
 	earmelfb_linux.c \
 	earmelfb_linux_eabi.c \
@@ -1064,6 +1064,7 @@
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/earmaoutl.Po@am__quote@
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/earmcoff.Po@am__quote@
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/earmelf.Po@am__quote@
+@AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/earmelf_aros.Po@am__quote@
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/earmelf_fbsd.Po@am__quote@
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/earmelf_linux.Po@am__quote@
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/earmelf_linux_eabi.Po@am__quote@
@@ -1072,6 +1073,7 @@
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/earmelf_nbsd.Po@am__quote@
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/earmelf_vxworks.Po@am__quote@
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/earmelfb.Po@am__quote@
+@AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/earmelfb_aros.Po@am__quote@
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/earmelfb_fbsd.Po@am__quote@
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/earmelfb_linux.Po@am__quote@
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/earmelfb_linux_eabi.Po@am__quote@
@@ -2158,6 +2158,12 @@
   $(ELF_DEPS) $(srcdir)/emultempl/armelf.em \
   $(srcdir)/scripttempl/elf.sc ${GEN_DEPENDS}
 
+earmelf_aros.c: $(srcdir)/emulparams/armelf_aros.sh \
+  $(srcdir)/emulparams/armelf_linux_eabi.sh \
+  $(srcdir)/emulparams/armelf_linux.sh \
+  $(ELF_DEPS) $(srcdir)/emultempl/armelf.em \
+  $(srcdir)/scripttempl/elf.sc ${GEN_DEPENDS}
+
 earmelf_nacl.c: $(srcdir)/emulparams/armelf_nacl.sh \
   $(srcdir)/emulparams/armelf_linux_eabi.sh \
   $(srcdir)/emulparams/armelf_linux.sh \
@@ -2196,6 +2196,13 @@
   $(ELF_DEPS) $(srcdir)/emultempl/armelf.em \
   $(srcdir)/scripttempl/elf.sc ${GEN_DEPENDS}
 
+earmelfb_aros.c: $(srcdir)/emulparams/armelfb_aros.sh \
+  $(srcdir)/emulparams/armelf_aros.sh \
+  $(srcdir)/emulparams/armelf_linux_eabi.sh \
+  $(srcdir)/emulparams/armelf_linux.sh \
+  $(ELF_DEPS) $(srcdir)/emultempl/armelf.em \
+  $(srcdir)/scripttempl/elf.sc ${GEN_DEPENDS}
+
 earmelfb_nacl.c: $(srcdir)/emulparams/armelfb_nacl.sh \
   $(srcdir)/emulparams/armelf_nacl.sh \
   $(srcdir)/emulparams/armelf_linux_eabi.sh \
