diff -ruN llvm-3.8.0.src/include/llvm/ADT/Triple.h llvm-3.8.0.src.aros/include/llvm/ADT/Triple.h
--- llvm-3.8.0.src/include/llvm/ADT/Triple.h	2015-12-01 10:33:56.000000000 +0000
+++ llvm-3.8.0.src.aros/include/llvm/ADT/Triple.h	2016-03-26 05:17:08.515284998 +0000
@@ -133,6 +133,7 @@
   enum OSType {
     UnknownOS,
 
+    AROS,
     CloudABI,
     Darwin,
     DragonFly,
@@ -403,6 +404,11 @@
     return isOSVersionLT(Minor + 4, Micro, 0);
   }
 
+  /// 
+    bool isAROS() const {
+    return getOS() == Triple::AROS;
+  }
+  
   /// isMacOSX - Is this a Mac OS X triple. For legacy reasons, we support both
   /// "darwin" and "osx" as OS X triples.
   bool isMacOSX() const {
diff -ruN llvm-3.8.0.src/lib/Support/Triple.cpp llvm-3.8.0.src.aros/lib/Support/Triple.cpp
--- llvm-3.8.0.src/lib/Support/Triple.cpp	2016-02-16 21:46:52.000000000 +0000
+++ llvm-3.8.0.src.aros/lib/Support/Triple.cpp	2016-03-26 05:20:23.835630566 +0000
@@ -157,6 +157,7 @@
   switch (Kind) {
   case UnknownOS: return "unknown";
 
+  case AROS: return "aros";
   case CloudABI: return "cloudabi";
   case Darwin: return "darwin";
   case DragonFly: return "dragonfly";
@@ -413,6 +414,7 @@
 
 static Triple::OSType parseOS(StringRef OSName) {
   return StringSwitch<Triple::OSType>(OSName)
+    .StartsWith("aros", Triple::AROS)
     .StartsWith("cloudabi", Triple::CloudABI)
     .StartsWith("darwin", Triple::Darwin)
     .StartsWith("dragonfly", Triple::DragonFly)
