diff -ur '--exclude-from=./mesa/src/aros/mesa.diff.excluded' -I '^.*\bRevision.*Date.*\b.*$' /data/deadwood/source/Mesa-7.11-staging/include/EGL/eglplatform.h ./mesa/include/EGL/eglplatform.h
--- /data/deadwood/source/Mesa-7.11-staging/include/EGL/eglplatform.h	2011-07-09 03:37:09.000000000 +0200
+++ ./mesa/include/EGL/eglplatform.h	2011-10-18 19:34:24.000000000 +0200
@@ -90,6 +90,14 @@
 typedef struct gbm_bo      *EGLNativePixmapType;
 typedef void               *EGLNativeWindowType;
 
+#elif defined(__AROS__)
+
+#include <intuition/intuition.h>
+
+typedef APTR            EGLNativeDisplayType;
+typedef struct Bitmap   *EGLNativePixmapType;
+typedef struct Window   *EGLNativeWindowType;
+
 #elif defined(__unix__) || defined(__unix)
 
 #ifdef MESA_EGL_NO_X11_HEADERS
@@ -127,6 +135,6 @@
  * handles are 64 bit types, then EGLint should be defined as a signed 64-bit
  * integer type.
  */
-typedef khronos_int32_t EGLint;
+typedef SIPTR EGLint;
 
 #endif /* __eglplatform_h */
diff -ur '--exclude-from=./mesa/src/aros/mesa.diff.excluded' -I '^.*\bRevision.*Date.*\b.*$' /data/deadwood/source/Mesa-7.11-staging/src/egl/main/eglcompiler.h ./mesa/src/egl/main/eglcompiler.h
--- /data/deadwood/source/Mesa-7.11-staging/src/egl/main/eglcompiler.h	2011-07-09 03:37:09.000000000 +0200
+++ ./mesa/src/egl/main/eglcompiler.h	2013-01-04 19:25:40.000000000 +0100
@@ -36,6 +36,7 @@
  */
 #if (defined(__STDC_VERSION__) && __STDC_VERSION__ >= 199901L)
 #  include <stdint.h>
+#  include <stddef.h>
 #elif defined(_MSC_VER)
    typedef __int8             int8_t;
    typedef unsigned __int8    uint8_t;
diff -ur '--exclude-from=./mesa/src/aros/mesa.diff.excluded' -I '^.*\bRevision.*Date.*\b.*$' /data/deadwood/source/Mesa-7.11-staging/src/egl/main/eglconfig.c ./mesa/src/egl/main/eglconfig.c
--- /data/deadwood/source/Mesa-7.11-staging/src/egl/main/eglconfig.c	2011-07-09 03:37:09.000000000 +0200
+++ ./mesa/src/egl/main/eglconfig.c	2013-01-04 19:25:40.000000000 +0100
@@ -263,7 +263,7 @@
 
    /* check attributes by their types */
    for (i = 0; i < ARRAY_SIZE(_eglValidationTable); i++) {
-      EGLint mask;
+      EGLint mask = 0;
 
       attr = _eglValidationTable[i].attr;
       val = _eglGetConfigKey(conf, attr);
@@ -700,7 +700,7 @@
 }
 
 
-static int
+static EGLint
 _eglFallbackCompare(const _EGLConfig *conf1, const _EGLConfig *conf2,
                    void *priv_data)
 {
diff -ur '--exclude-from=./mesa/src/aros/mesa.diff.excluded' -I '^.*\bRevision.*Date.*\b.*$' /data/deadwood/source/Mesa-7.11-staging/src/egl/main/eglcurrent.c ./mesa/src/egl/main/eglcurrent.c
--- /data/deadwood/source/Mesa-7.11-staging/src/egl/main/eglcurrent.c	2011-07-09 03:37:09.000000000 +0200
+++ ./mesa/src/egl/main/eglcurrent.c	2011-09-05 20:03:01.000000000 +0200
@@ -108,6 +108,55 @@
    return EGL_TRUE;
 }
 
+#elif defined(_EGL_OS_AROS)
+
+#include "aros/tls.h"
+
+static struct TaskLocalStorage * tls = NULL;
+static void (*_egl_FreeTSD)(_EGLThreadInfo *);
+
+static INLINE void _eglFiniTSD(void)
+{
+    DestroyTLS(tls);
+}
+
+static INLINE void _eglSetTSD(const _EGLThreadInfo *t)
+{
+    InsertIntoTLS(tls, (APTR)t);
+}
+
+static INLINE _EGLThreadInfo *_eglGetTSD(void)
+{
+    return (_EGLThreadInfo *)GetFromTLS(tls);
+}
+
+static INLINE EGLBoolean _eglInitTSD(void (*dtor)(_EGLThreadInfo *))
+{
+    if (!tls)
+    {
+        tls = CreateTLS();
+        _egl_FreeTSD = dtor;
+        _eglAddAtExitCall(_eglFiniTSD);
+    }
+
+    if (tls)
+        return EGL_TRUE;
+    else
+        return EGL_FALSE;
+}
+
+#include <aros/symbolsets.h>
+
+static VOID _egl_FreeTSD_fn()
+{
+    _EGLThreadInfo *t = _eglGetTSD();
+
+    if (t && _egl_FreeTSD)
+        _egl_FreeTSD((void *) t);
+    ClearFromTLS(tls);
+}
+ADD2CLOSELIB(_egl_FreeTSD_fn, 10)
+
 #else /* PTHREADS */
 static const _EGLThreadInfo *_egl_TSD;
 static void (*_egl_FreeTSD)(_EGLThreadInfo *);
diff -ur '--exclude-from=./mesa/src/aros/mesa.diff.excluded' -I '^.*\bRevision.*Date.*\b.*$' /data/deadwood/source/Mesa-7.11-staging/src/egl/main/egldisplay.c ./mesa/src/egl/main/egldisplay.c
--- /data/deadwood/source/Mesa-7.11-staging/src/egl/main/egldisplay.c	2011-07-09 03:37:09.000000000 +0200
+++ ./mesa/src/egl/main/egldisplay.c	2011-09-05 20:03:01.000000000 +0200
@@ -59,7 +59,8 @@
       { _EGL_PLATFORM_X11, "x11" },
       { _EGL_PLATFORM_WAYLAND, "wayland" },
       { _EGL_PLATFORM_DRM, "drm" },
-      { _EGL_PLATFORM_FBDEV, "fbdev" }
+      { _EGL_PLATFORM_FBDEV, "fbdev" },
+      { _EGL_PLATFORM_AROS, "aros" }
    };
    _EGLPlatformType plat = _EGL_INVALID_PLATFORM;
    const char *plat_name;
@@ -148,8 +149,15 @@
    /* search the display list first */
    dpy = _eglGlobal.DisplayList;
    while (dpy) {
+#if !defined(_EGL_OS_AROS)
+   /* It seems the _eglGlobal.DisplayList should unique per opener (not per task).
+      This is not true on AROS (_eglGlobal.DisplayList  is global for all 
+      openers). The workaround is to always create a new display object. This 
+      might fail with multithreaded applications as they might expect to have 
+      the same display object returned by GetDisplay */
       if (dpy->Platform == plat && dpy->PlatformDisplay == plat_dpy)
          break;
+#endif
       dpy = dpy->Next;
    }
 
diff -ur '--exclude-from=./mesa/src/aros/mesa.diff.excluded' -I '^.*\bRevision.*Date.*\b.*$' /data/deadwood/source/Mesa-7.11-staging/src/egl/main/egldisplay.h ./mesa/src/egl/main/egldisplay.h
--- /data/deadwood/source/Mesa-7.11-staging/src/egl/main/egldisplay.h	2011-07-09 03:37:09.000000000 +0200
+++ ./mesa/src/egl/main/egldisplay.h	2011-09-05 20:03:01.000000000 +0200
@@ -44,6 +44,7 @@
    _EGL_PLATFORM_WAYLAND,
    _EGL_PLATFORM_DRM,
    _EGL_PLATFORM_FBDEV,
+   _EGL_PLATFORM_AROS,
 
    _EGL_NUM_PLATFORMS,
    _EGL_INVALID_PLATFORM = -1
diff -ur '--exclude-from=./mesa/src/aros/mesa.diff.excluded' -I '^.*\bRevision.*Date.*\b.*$' /data/deadwood/source/Mesa-7.11-staging/src/egl/main/egldriver.c ./mesa/src/egl/main/egldriver.c
--- /data/deadwood/source/Mesa-7.11-staging/src/egl/main/egldriver.c	2011-07-09 03:37:09.000000000 +0200
+++ ./mesa/src/egl/main/egldriver.c	2012-01-22 09:08:51.000000000 +0100
@@ -32,7 +32,6 @@
  * Functions for choosing and opening/loading device drivers.
  */
 
-
 #include <assert.h>
 #include <string.h>
 #include <stdio.h>
@@ -79,6 +78,7 @@
    { NULL, NULL }
 };
 
+
 /**
  * Wrappers for dlopen/dlclose()
  */
@@ -107,6 +107,16 @@
 }
 
 
+#elif defined(_EGL_OS_AROS)
+
+typedef void * lib_handle;
+
+static void
+close_library(void *lib)
+{
+    /* No-Op */
+}
+
 #elif defined(_EGL_OS_UNIX)
 
 
@@ -141,6 +151,7 @@
 static _EGLMain_t
 _eglOpenLibrary(const char *driverPath, lib_handle *handle)
 {
+#if !defined(_EGL_OS_AROS)
    lib_handle lib;
    _EGLMain_t mainFunc = NULL;
    const char *error = "unknown error";
@@ -187,6 +198,9 @@
 
    *handle = lib;
    return mainFunc;
+#else
+   return NULL;
+#endif
 }
 
 
@@ -314,6 +328,7 @@
 }
 
 
+#if !defined(_EGL_OS_AROS)
 /**
  * A loader function for use with _eglPreloadForEach.  The loader data is the
  * filename of the driver.   This function stops on the first valid driver.
@@ -531,7 +546,7 @@
    _eglPreloadForEach(_eglGetSearchPath(), _eglLoaderFile, external);
 #endif
 }
-
+#endif /* !defined(_EGL_OS_AROS) */
 
 /**
  * Add built-in drivers to the module array.
@@ -560,6 +575,7 @@
    if (_eglModules)
       return EGL_TRUE;
 
+#if !defined(_EGL_OS_AROS)
    if (!_eglAddUserDriver()) {
       /*
        * Add other drivers only when EGL_DRIVER is not set.  The order here
@@ -568,6 +584,9 @@
       _eglAddGalliumDriver();
       _eglAddBuiltInDrivers();
    }
+#else
+   _eglAddBuiltInDrivers();
+#endif
 
    return (_eglModules != NULL);
 }
@@ -698,7 +717,7 @@
    }
 }
 
-
+#if !defined(_EGL_OS_AROS)
 /**
  * Invoke a callback function on each EGL search path.
  *
@@ -712,3 +731,4 @@
    const char *search_path = _eglGetSearchPath();
    _eglPreloadForEach(search_path, callback, callback_data);
 }
+#endif
diff -ur '--exclude-from=./mesa/src/aros/mesa.diff.excluded' -I '^.*\bRevision.*Date.*\b.*$' /data/deadwood/source/Mesa-7.11-staging/src/egl/main/egldriver.h ./mesa/src/egl/main/egldriver.h
--- /data/deadwood/source/Mesa-7.11-staging/src/egl/main/egldriver.h	2011-07-09 03:37:09.000000000 +0200
+++ ./mesa/src/egl/main/egldriver.h	2012-01-21 16:25:55.000000000 +0100
@@ -95,6 +95,10 @@
 extern _EGLDriver *
 _eglBuiltInDriverGLX(const char *args);
 
+#if defined(_EGL_OS_AROS)
+extern _EGLDriver *
+_eglBuiltInDriverAROSMesa(const char *args);
+#endif
 
 PUBLIC _EGLDriver *
 _eglMain(const char *args);
diff -ur '--exclude-from=./mesa/src/aros/mesa.diff.excluded' -I '^.*\bRevision.*Date.*\b.*$' /data/deadwood/source/Mesa-7.11-staging/src/egl/main/egllog.c ./mesa/src/egl/main/egllog.c
--- /data/deadwood/source/Mesa-7.11-staging/src/egl/main/egllog.c	2011-07-09 03:37:09.000000000 +0200
+++ ./mesa/src/egl/main/egllog.c	2011-09-05 20:03:01.000000000 +0200
@@ -56,11 +56,18 @@
    _EGLLogProc logger;
    EGLint num_messages;
 } logging = {
+#if !defined(_EGL_OS_AROS)
    _EGL_MUTEX_INITIALIZER,
    EGL_FALSE,
    FALLBACK_LOG_LEVEL,
    NULL,
    0
+#else
+   .initialized = EGL_FALSE,
+   .level = FALLBACK_LOG_LEVEL,
+   .logger = NULL,
+   .num_messages = 0
+#endif
 };
 
 static const char *level_strings[] = {
@@ -122,13 +129,20 @@
 }
 
 
+#if defined(_EGL_OS_AROS)
+#include <aros/debug.h>
+#endif
 /**
  * The default logger.  It prints the message to stderr.
  */
 static void
 _eglDefaultLogger(EGLint level, const char *msg)
 {
+#if !defined(_EGL_OS_AROS)
    fprintf(stderr, "libEGL %s: %s\n", level_strings[level], msg);
+#else
+   bug("[EGL]: %s: %s\n", level_strings[level], msg);
+#endif
 }
 
 
@@ -160,6 +174,9 @@
    logging.logger = _eglDefaultLogger;
    logging.level = (level >= 0) ? level : FALLBACK_LOG_LEVEL;
    logging.initialized = EGL_TRUE;
+#if defined(_EGL_OS_AROS)
+   _eglInitMutex(&logging.mutex);
+#endif
 
    /* it is fine to call _eglLog now */
    if (log_env && level < 0) {
diff -ur '--exclude-from=./mesa/src/aros/mesa.diff.excluded' -I '^.*\bRevision.*Date.*\b.*$' /data/deadwood/source/Mesa-7.11-staging/src/egl/main/eglmutex.h ./mesa/src/egl/main/eglmutex.h
--- /data/deadwood/source/Mesa-7.11-staging/src/egl/main/eglmutex.h	2011-07-09 03:37:09.000000000 +0200
+++ ./mesa/src/egl/main/eglmutex.h	2011-09-05 20:03:01.000000000 +0200
@@ -63,6 +63,48 @@
 #define _EGL_DECLARE_MUTEX(m) \
    _EGLMutex m = _EGL_MUTEX_INITIALIZER
 
+#elif defined(_EGL_OS_AROS)
+
+#include <proto/exec.h>
+
+typedef struct SignalSemaphore _EGLMutex;
+
+static INLINE void _eglInitMutex(_EGLMutex *m)
+{
+   InitSemaphore(m);
+}
+
+static INLINE void
+_eglDestroyMutex(_EGLMutex *m)
+{
+   /* No Op */
+}
+
+static INLINE void
+_eglLockMutex(_EGLMutex *m)
+{
+   ObtainSemaphore(m);
+}
+
+static INLINE void
+_eglUnlockMutex(_EGLMutex *m)
+{
+   ReleaseSemaphore(m);
+}
+
+/* This is not supported. _eglInitMutex must alway be used */
+/* #define _EGL_MUTEX_INITIALIZER */
+
+#include <aros/symbolsets.h> /* For ADD2INIT */
+/* Declare variable, declare init function, add to auto init. Ugly but works. */
+#define _EGL_DECLARE_MUTEX(m)   \
+   _EGLMutex m;                 \
+static void init##m()           \
+{                               \
+    _eglInitMutex(&m);          \
+}                               \
+ADD2INIT(init##m, 5);
+
 #else
 
 typedef int _EGLMutex;
diff -ur '--exclude-from=./mesa/src/aros/mesa.diff.excluded' -I '^.*\bRevision.*Date.*\b.*$' /data/deadwood/source/Mesa-7.11-staging/src/gallium/auxiliary/os/os_thread.h ./mesa/src/gallium/auxiliary/os/os_thread.h
--- /data/deadwood/source/Mesa-7.11-staging/src/gallium/auxiliary/os/os_thread.h	2011-07-09 03:37:09.000000000 +0200
+++ ./mesa/src/gallium/auxiliary/os/os_thread.h	2011-09-05 20:03:01.000000000 +0200
@@ -237,6 +237,63 @@
 
 #endif /* pre-Vista win32 */
 
+#elif defined(PIPE_OS_AROS)
+
+#include <proto/exec.h>
+
+#include "os/os_time.h"
+
+typedef struct SignalSemaphore pipe_mutex;
+
+#include <aros/symbolsets.h> /* For ADD2INIT */
+
+/* Declare variable, declare init function, add to auto init. Ugly but works. */
+#define pipe_static_mutex(mutex) \
+static pipe_mutex mutex;         \
+static void init##mutex()        \
+{                                \
+    pipe_mutex_init(mutex);      \
+}                                \
+ADD2INIT(init##mutex, 5);
+
+#define pipe_mutex_init(mutex) \
+   InitSemaphore(&mutex) 
+
+#define pipe_mutex_destroy(mutex) \
+   (void) mutex
+
+#define pipe_mutex_lock(mutex) \
+   ObtainSemaphore(&mutex)
+
+#define pipe_mutex_unlock(mutex) \
+   ReleaseSemaphore(&mutex)
+
+typedef int64_t pipe_condvar;
+
+#define pipe_static_condvar(condvar) \
+   static pipe_condvar condvar = 1000
+
+#define pipe_condvar_init(condvar) \
+   (void) (condvar = 1000)
+
+#define pipe_condvar_destroy(condvar) \
+   (void) condvar
+
+/* Poor man's pthread_cond_wait():
+   Just release the mutex and sleep for one millisecond.
+   The caller's while() loop does all the work. */
+#define pipe_condvar_wait(condvar, mutex) \
+   do { pipe_mutex_unlock(mutex); \
+        os_time_sleep(condvar); \
+        pipe_mutex_lock(mutex); \
+   } while (0)
+
+#define pipe_condvar_signal(condvar) \
+   (void) condvar
+
+#define pipe_condvar_broadcast(condvar) \
+   (void) condvar
+
 #else
 
 #include "os/os_time.h"
diff -ur '--exclude-from=./mesa/src/aros/mesa.diff.excluded' -I '^.*\bRevision.*Date.*\b.*$' /data/deadwood/source/Mesa-7.11-staging/src/gallium/auxiliary/os/os_time.c ./mesa/src/gallium/auxiliary/os/os_time.c
--- /data/deadwood/source/Mesa-7.11-staging/src/gallium/auxiliary/os/os_time.c	2011-07-09 03:37:09.000000000 +0200
+++ ./mesa/src/gallium/auxiliary/os/os_time.c	2011-09-05 20:03:01.000000000 +0200
@@ -35,7 +35,7 @@
 
 #include "pipe/p_config.h"
 
-#if defined(PIPE_OS_UNIX)
+#if defined(PIPE_OS_UNIX) || defined(PIPE_OS_AROS)
 #  include <sys/time.h> /* timeval */
 #elif defined(PIPE_SUBSYSTEM_WINDOWS_DISPLAY)
 #  include <windows.h>
@@ -55,7 +55,7 @@
 int64_t
 os_time_get(void)
 {
-#if defined(PIPE_OS_UNIX)
+#if defined(PIPE_OS_UNIX) || defined(PIPE_OS_AROS)
 
    struct timeval tv;
    gettimeofday(&tv, NULL);
diff -ur '--exclude-from=./mesa/src/aros/mesa.diff.excluded' -I '^.*\bRevision.*Date.*\b.*$' /data/deadwood/source/Mesa-7.11-staging/src/gallium/auxiliary/os/os_time.h ./mesa/src/gallium/auxiliary/os/os_time.h
--- /data/deadwood/source/Mesa-7.11-staging/src/gallium/auxiliary/os/os_time.h	2010-04-27 23:41:59.000000000 +0200
+++ ./mesa/src/gallium/auxiliary/os/os_time.h	2011-08-31 12:37:34.000000000 +0200
@@ -38,7 +38,7 @@
 
 #include "pipe/p_config.h"
 
-#if defined(PIPE_OS_UNIX)
+#if defined(PIPE_OS_UNIX) || defined(PIPE_OS_AROS)
 #  include <unistd.h> /* usleep */
 #endif
 
@@ -60,7 +60,7 @@
 /*
  * Sleep.
  */
-#if defined(PIPE_OS_UNIX)
+#if defined(PIPE_OS_UNIX) || defined(PIPE_OS_AROS)
 #define os_time_sleep(_usecs) usleep(_usecs)
 #else
 void
diff -ur '--exclude-from=./mesa/src/aros/mesa.diff.excluded' -I '^.*\bRevision.*Date.*\b.*$' /data/deadwood/source/Mesa-7.11-staging/src/gallium/auxiliary/util/u_atomic.h ./mesa/src/gallium/auxiliary/util/u_atomic.h
--- /data/deadwood/source/Mesa-7.11-staging/src/gallium/auxiliary/util/u_atomic.h	2010-10-19 19:58:28.000000000 +0200
+++ ./mesa/src/gallium/auxiliary/util/u_atomic.h	2011-12-17 11:25:59.000000000 +0100
@@ -29,8 +29,12 @@
 #define PIPE_ATOMIC_ASM_MSVC_X86                
 #elif (defined(PIPE_CC_GCC) && defined(PIPE_ARCH_X86))
 #define PIPE_ATOMIC_ASM_GCC_X86
+#elif (defined(PIPE_CC_GCC) && defined(PIPE_ARCH_ARM))
+#define PIPE_ATOMIC_ASM_GCC_ARM
 #elif (defined(PIPE_CC_GCC) && defined(PIPE_ARCH_X86_64))
 #define PIPE_ATOMIC_ASM_GCC_X86_64
+#elif defined(PIPE_OS_AROS) && defined(PIPE_ARCH_M68K)
+#define PIPE_ATOMIC_OS_AROS_CPU_M68K
 #elif defined(PIPE_CC_GCC) && (PIPE_CC_GCC_VERSION >= 401)
 #define PIPE_ATOMIC_GCC_INTRINSIC
 #else
@@ -130,6 +134,69 @@
 
 #endif
 
+#if defined(PIPE_ATOMIC_ASM_GCC_ARM)
+
+#define PIPE_ATOMIC "GCC ARM assembly"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#define p_atomic_set(_v, _i) (*(_v) = (_i))
+#define p_atomic_read(_v) (*(_v))
+
+static INLINE boolean
+p_atomic_dec_zero(int32_t *v)
+{
+   unsigned long temp;
+   int result;
+   unsigned long cc;
+   __asm__ __volatile__("\n1: ldrex %0, [%3]; subs %0, %0, #1; moveq %2, #1; movne %2, #0; strex %1, %0, [%3]; teq %1, #0; bne 1b"
+						   :"=&r"(result), "=&r"(temp), "=&r"(cc)
+						   :"r"(v)
+						   :"cc");
+   return cc;
+}
+
+static INLINE void
+p_atomic_inc(int32_t *v)
+{
+   unsigned long temp;
+   int result;
+   __asm__ __volatile__("\n1: ldrex %0, [%2]; add %0, %0, #1; strex %1, %0, [%2]; teq %1, #0; bne 1b"
+		   	   	   	   	   :"=&r"(result), "=&r"(temp)
+		   	   	   	   	   :"r"(v)
+		   	   	   	   	   :"cc");
+}
+
+static INLINE void
+p_atomic_dec(int32_t *v)
+{
+   unsigned long temp;
+   int result;
+   __asm__ __volatile__("\n1: ldrex %0, [%2]; sub %0, %0, #1; strex %1, %0, [%2]; teq %1, #0; bne 1b"
+			   	   	   	   :"=&r"(result), "=&r"(temp)
+			   	   	   	   :"r"(v)
+			   	   	   	   :"cc");
+}
+
+static INLINE int32_t
+p_atomic_cmpxchg(int32_t *v, int32_t old, int32_t _new)
+{
+	  int32_t oldval;
+	   unsigned long temp;
+	   __asm__ __volatile__("\n1: ldrex %0,[%2]; teq %0, %3; strexeq %1, %4, [%2]; teq %1, #0; bne 1b"
+	                                                   :"=&r"(oldval), "=&r"(temp)
+	                                                   :"r"(v), "Ir"(old), "r"(_new)
+	                                                   :"cc");
+	   return oldval;
+}
+
+#ifdef __cplusplus
+}
+#endif
+
+#endif
 
 
 /* Implementation using GCC-provided synchronization intrinsics
@@ -338,6 +405,64 @@
 
 #ifdef __cplusplus
 }
+#endif
+
+#endif
+
+
+#if defined(PIPE_ATOMIC_OS_AROS_CPU_M68K)
+
+#define PIPE_ATOMIC "AROS OS atomic functions"
+
+#include <aros/atomic.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#define p_atomic_set(_v, _i) (*(_v) = (_i))
+#define p_atomic_read(_v) (*(_v))
+
+static INLINE boolean
+p_atomic_dec_zero(int32_t *v)
+{
+   boolean n;
+ 
+   /* FIXME: AROS needs an atomic decrement and return... */
+   Disable();
+   AROS_ATOMIC_DEC(*(LONG *)v);
+   n = (*v == 0) ? TRUE : FALSE;
+   Enable();
+
+   return n;
+}
+
+#define p_atomic_inc(_v) AROS_ATOMIC_INC(*(LONG *)_v)
+#define p_atomic_dec(_v) AROS_ATOMIC_DEC(*(LONG *)_v)
+
+static INLINE int32_t
+p_atomic_cmpxchg(int32_t *v, int32_t o, int32_t n)
+{
+	int32_t ret;
+
+	/* FIXME: AROS needs an atomic cmpxchg, using CAS.
+	 * However we can't do this if:
+	 *  a) We are on a 68000 or
+	 *  b) The 'v' points to Chip RAM (no r/m/w possible)
+	 *
+	 *  Settle for Disable()/Enable() for now.
+	 */
+	Disable();
+	if (*v == o)
+		*v = (n);
+	ret = *v;
+	Enable();
+
+	return ret;
+}
+
+#ifdef __cplusplus
+}
 #endif
 
 #endif
diff -ur '--exclude-from=./mesa/src/aros/mesa.diff.excluded' -I '^.*\bRevision.*Date.*\b.*$' /data/deadwood/source/Mesa-7.11-staging/src/gallium/auxiliary/util/u_math.c ./mesa/src/gallium/auxiliary/util/u_math.c
--- /data/deadwood/source/Mesa-7.11-staging/src/gallium/auxiliary/util/u_math.c	2010-02-05 01:10:39.000000000 +0100
+++ ./mesa/src/gallium/auxiliary/util/u_math.c	2011-08-31 12:37:34.000000000 +0200
@@ -42,6 +42,13 @@
       pow2_table[i] = (float) pow(2.0, (i - POW2_TABLE_OFFSET) / POW2_TABLE_SCALE);
 }
 
+#if defined(PIPE_OS_AROS)
+/*
+ * NOTE: log_base_2(x) = log(x) / log(2)
+ * NOTE: 1.442695 = 1/log(2).
+ */
+#define log2(x)  ((float) (log(x) * 1.442695f))
+#endif
 
 /** log2(x), for x in [1.0, 2.0) */
 float log2_table[LOG2_TABLE_SIZE];
diff -ur '--exclude-from=./mesa/src/aros/mesa.diff.excluded' -I '^.*\bRevision.*Date.*\b.*$' /data/deadwood/source/Mesa-7.11-staging/src/gallium/auxiliary/util/u_math.h ./mesa/src/gallium/auxiliary/util/u_math.h
--- /data/deadwood/source/Mesa-7.11-staging/src/gallium/auxiliary/util/u_math.h	2011-07-09 03:37:09.000000000 +0200
+++ ./mesa/src/gallium/auxiliary/util/u_math.h	2011-09-05 20:03:01.000000000 +0200
@@ -411,6 +411,8 @@
 }
 #elif defined(__MINGW32__)
 #define ffs __builtin_ffs
+#elif defined(PIPE_OS_AROS)
+#define ffs __builtin_ffs
 #endif
 
 
@@ -530,7 +532,15 @@
 util_bitcount(unsigned n)
 {
 #if defined(PIPE_CC_GCC) && (PIPE_CC_GCC_VERSION >= 304)
+#if defined(PIPE_OS_AROS)
+   unsigned int bits;
+   for (bits = 0; n > 0; n = n >> 1) {
+      bits += (n & 1);
+   }
+   return bits;
+#else
    return __builtin_popcount(n);
+#endif
 #else
    /* K&R classic bitcount.
     *
diff -ur '--exclude-from=./mesa/src/aros/mesa.diff.excluded' -I '^.*\bRevision.*Date.*\b.*$' /data/deadwood/source/Mesa-7.11-staging/src/gallium/auxiliary/util/u_network.c ./mesa/src/gallium/auxiliary/util/u_network.c
--- /data/deadwood/source/Mesa-7.11-staging/src/gallium/auxiliary/util/u_network.c	2010-10-05 18:56:51.000000000 +0200
+++ ./mesa/src/gallium/auxiliary/util/u_network.c	2013-03-13 22:19:46.000000000 +0100
@@ -6,7 +6,7 @@
 #if defined(PIPE_SUBSYSTEM_WINDOWS_USER)
 #  include <winsock2.h>
 #  include <windows.h>
-#elif defined(PIPE_OS_LINUX) || defined(PIPE_OS_HAIKU) || defined(PIPE_OS_APPLE) || defined(PIPE_OS_CYGWIN)
+#elif defined(PIPE_OS_LINUX) || defined(PIPE_OS_HAIKU) || defined(PIPE_OS_APPLE) || defined(PIPE_OS_CYGWIN) || defined(PIPE_OS_AROS)
 #  include <sys/socket.h>
 #  include <netinet/in.h>
 #  include <unistd.h>
diff -ur '--exclude-from=./mesa/src/aros/mesa.diff.excluded' -I '^.*\bRevision.*Date.*\b.*$' /data/deadwood/source/Mesa-7.11-staging/src/gallium/drivers/i915/i915_context.c ./mesa/src/gallium/drivers/i915/i915_context.c
--- /data/deadwood/source/Mesa-7.11-staging/src/gallium/drivers/i915/i915_context.c	2011-07-09 03:37:09.000000000 +0200
+++ ./mesa/src/gallium/drivers/i915/i915_context.c	2013-03-13 22:19:46.000000000 +0100
@@ -54,13 +54,10 @@
    struct i915_context *i915 = i915_context(pipe);
    struct draw_context *draw = i915->draw;
    void *mapped_indices = NULL;
-   unsigned cbuf_dirty;
-
 
    /*
     * Ack vs contants here, helps ipers a lot.
     */
-   cbuf_dirty = i915->dirty & I915_NEW_VS_CONSTANTS;
    i915->dirty &= ~I915_NEW_VS_CONSTANTS;
 
    if (i915->dirty)
diff -ur '--exclude-from=./mesa/src/aros/mesa.diff.excluded' -I '^.*\bRevision.*Date.*\b.*$' /data/deadwood/source/Mesa-7.11-staging/src/gallium/drivers/i915/i915_prim_vbuf.c ./mesa/src/gallium/drivers/i915/i915_prim_vbuf.c
--- /data/deadwood/source/Mesa-7.11-staging/src/gallium/drivers/i915/i915_prim_vbuf.c	2011-07-09 03:37:09.000000000 +0200
+++ ./mesa/src/gallium/drivers/i915/i915_prim_vbuf.c	2013-01-04 19:25:40.000000000 +0100
@@ -400,8 +400,8 @@
    case PIPE_PRIM_LINE_LOOP:
       if (nr >= 2) {
          for (i = start + 1; i < end; i++)
-            OUT_BATCH((i-0) | (i+0) << 16);
-         OUT_BATCH((i-0) | ( start) << 16);
+            OUT_BATCH((i-1) | (i+0) << 16);
+         OUT_BATCH((i-1) | ( start) << 16);
       }
       break;
    case PIPE_PRIM_QUADS:
diff -ur '--exclude-from=./mesa/src/aros/mesa.diff.excluded' -I '^.*\bRevision.*Date.*\b.*$' /data/deadwood/source/Mesa-7.11-staging/src/gallium/drivers/i915/i915_screen.c ./mesa/src/gallium/drivers/i915/i915_screen.c
--- /data/deadwood/source/Mesa-7.11-staging/src/gallium/drivers/i915/i915_screen.c	2011-07-09 03:37:09.000000000 +0200
+++ ./mesa/src/gallium/drivers/i915/i915_screen.c	2013-01-04 19:25:40.000000000 +0100
@@ -274,6 +274,12 @@
       PIPE_FORMAT_YUYV,
       /* XXX why not?
       PIPE_FORMAT_Z16_UNORM, */
+      
+      // AROS Quoke1 fix,seems to work,but probably better to fix SDL so
+      // that some other format is used.
+      // Sami
+      PIPE_FORMAT_Z16_UNORM,
+      
       PIPE_FORMAT_DXT1_RGB,
       PIPE_FORMAT_DXT1_RGBA,
       PIPE_FORMAT_DXT3_RGBA,
@@ -294,6 +300,7 @@
    static const enum pipe_format depth_supported[] = {
       /* XXX why not?
       PIPE_FORMAT_Z16_UNORM, */
+      PIPE_FORMAT_Z16_UNORM, // AROS quake1 fix
       PIPE_FORMAT_Z24X8_UNORM,
       PIPE_FORMAT_Z24_UNORM_S8_USCALED,
       PIPE_FORMAT_NONE  /* list terminator */
diff -ur '--exclude-from=./mesa/src/aros/mesa.diff.excluded' -I '^.*\bRevision.*Date.*\b.*$' /data/deadwood/source/Mesa-7.11-staging/src/gallium/drivers/i915/i915_state_immediate.c ./mesa/src/gallium/drivers/i915/i915_state_immediate.c
--- /data/deadwood/source/Mesa-7.11-staging/src/gallium/drivers/i915/i915_state_immediate.c	2011-07-09 03:32:30.000000000 +0200
+++ ./mesa/src/gallium/drivers/i915/i915_state_immediate.c	2013-03-13 22:19:46.000000000 +0100
@@ -189,13 +189,13 @@
  */
 static void upload_S7(struct i915_context *i915)
 {
+#if 0
    unsigned LIS7;
 
    /* I915_NEW_RASTERIZER
     */
    LIS7 = i915->rasterizer->LIS7;
 
-#if 0
    set_immediate(i915, I915_IMMEDIATE_S7, LIS7);
 #endif
 }
diff -ur '--exclude-from=./mesa/src/aros/mesa.diff.excluded' -I '^.*\bRevision.*Date.*\b.*$' /data/deadwood/source/Mesa-7.11-staging/src/gallium/drivers/nouveau/nouveau_fence.c ./mesa/src/gallium/drivers/nouveau/nouveau_fence.c
--- /data/deadwood/source/Mesa-7.11-staging/src/gallium/drivers/nouveau/nouveau_fence.c	2011-07-09 03:37:09.000000000 +0200
+++ ./mesa/src/gallium/drivers/nouveau/nouveau_fence.c	2013-01-04 19:25:40.000000000 +0100
@@ -31,6 +31,10 @@
 #include <sched.h>
 #endif
 
+#ifdef PIPE_OS_AROS
+#include "drm_aros_config.h" /* for MOCK_HARDWARE define */
+#endif
+
 boolean
 nouveau_fence_new(struct nouveau_screen *screen, struct nouveau_fence **fence,
                   boolean emit)
@@ -93,8 +97,6 @@
    /* set this now, so that if fence.emit triggers a flush we don't recurse */
    fence->state = NOUVEAU_FENCE_STATE_EMITTED;
 
-   screen->fence.emit(&screen->base, fence->sequence);
-
    ++fence->ref;
 
    if (screen->fence.tail)
@@ -103,6 +105,8 @@
       screen->fence.head = fence;
 
    screen->fence.tail = fence;
+
+   screen->fence.emit(&screen->base, fence->sequence);
 }
 
 void
@@ -138,7 +142,13 @@
 {
    struct nouveau_fence *fence;
    struct nouveau_fence *next = NULL;
+
+#if defined(PIPE_OS_AROS) && defined(MOCK_HARDWARE)
+   /* For purpose of simulation, assume all fences are signalled */
+   u32 sequence = screen->fence.tail->sequence;
+#else
    u32 sequence = screen->fence.update(&screen->base);
+#endif
 
    if (screen->fence.sequence_ack == sequence)
       return;
@@ -219,5 +229,7 @@
    if (screen->fence.current->state < NOUVEAU_FENCE_STATE_EMITTED)
       nouveau_fence_emit(screen->fence.current);
 
+   nouveau_fence_ref(NULL, &screen->fence.current);
+
    nouveau_fence_new(screen, &screen->fence.current, FALSE);
 }
diff -ur '--exclude-from=./mesa/src/aros/mesa.diff.excluded' -I '^.*\bRevision.*Date.*\b.*$' /data/deadwood/source/Mesa-7.11-staging/src/gallium/drivers/nv50/nv50_tgsi_to_nc.c ./mesa/src/gallium/drivers/nv50/nv50_tgsi_to_nc.c
--- /data/deadwood/source/Mesa-7.11-staging/src/gallium/drivers/nv50/nv50_tgsi_to_nc.c	2011-07-09 03:37:09.000000000 +0200
+++ ./mesa/src/gallium/drivers/nv50/nv50_tgsi_to_nc.c	2011-09-05 20:03:01.000000000 +0200
@@ -1480,7 +1480,7 @@
 {
    struct nv_value *t[4], *s[3];
    uint opcode = translate_opcode(insn->Instruction.Opcode);
-   int arg, dim, c;
+   int arg = 0, dim = 0, c;
    const int tic = insn->Src[1].Register.Index;
    const int tsc = tic;
    const int cube = (insn->Texture.Texture  == TGSI_TEXTURE_CUBE) ? 1 : 0;
diff -ur '--exclude-from=./mesa/src/aros/mesa.diff.excluded' -I '^.*\bRevision.*Date.*\b.*$' /data/deadwood/source/Mesa-7.11-staging/src/gallium/drivers/nvfx/nv30_fragtex.c ./mesa/src/gallium/drivers/nvfx/nv30_fragtex.c
--- /data/deadwood/source/Mesa-7.11-staging/src/gallium/drivers/nvfx/nv30_fragtex.c	2011-04-08 21:31:18.000000000 +0200
+++ ./mesa/src/gallium/drivers/nvfx/nv30_fragtex.c	2011-09-30 21:45:16.000000000 +0200
@@ -9,7 +9,9 @@
 			  struct nvfx_sampler_state *ps,
 			  const struct pipe_sampler_state *cso)
 {
+#if 0 /* unused */
 	float limit;
+#endif
 
 	if (cso->max_anisotropy >= 2)
 	{
@@ -21,7 +23,9 @@
 			ps->en |= NV30_3D_TEX_ENABLE_ANISO_2X;
 	}
 
+#if 0 /* unused */
 	limit = CLAMP(cso->lod_bias, -16.0, 15.0 + (255.0 / 256.0));
+#endif
 	ps->filt |= (int)(cso->lod_bias * 256.0) & 0x1fff;
 
 	ps->max_lod = (int)CLAMP(cso->max_lod, 0.0, 15.0);
diff -ur '--exclude-from=./mesa/src/aros/mesa.diff.excluded' -I '^.*\bRevision.*Date.*\b.*$' /data/deadwood/source/Mesa-7.11-staging/src/gallium/drivers/nvfx/nv40_fragtex.c ./mesa/src/gallium/drivers/nvfx/nv40_fragtex.c
--- /data/deadwood/source/Mesa-7.11-staging/src/gallium/drivers/nvfx/nv40_fragtex.c	2011-04-08 21:31:18.000000000 +0200
+++ ./mesa/src/gallium/drivers/nvfx/nv40_fragtex.c	2011-09-30 21:45:16.000000000 +0200
@@ -8,7 +8,9 @@
 			  struct nvfx_sampler_state *ps,
 			  const struct pipe_sampler_state *cso)
 {
+#if 0 /* unused */
 	float limit;
+#endif
 	if (cso->max_anisotropy >= 2) {
 		/* no idea, binary driver sets it, works without it.. meh.. */
 		ps->wrap |= (1 << 5);
@@ -29,7 +31,9 @@
 			ps->en |= NV40_3D_TEX_ENABLE_ANISO_2X;
 	}
 
+#if 0 /* unused */
 	limit = CLAMP(cso->lod_bias, -16.0, 15.0 + (255.0 / 256.0));
+#endif
 	ps->filt |= (int)(cso->lod_bias * 256.0) & 0x1fff;
 
 	ps->max_lod = (int)(CLAMP(cso->max_lod, 0.0, 15.0 + (255.0 / 256.0)) * 256.0);
diff -ur '--exclude-from=./mesa/src/aros/mesa.diff.excluded' -I '^.*\bRevision.*Date.*\b.*$' /data/deadwood/source/Mesa-7.11-staging/src/gallium/drivers/softpipe/sp_query.c ./mesa/src/gallium/drivers/softpipe/sp_query.c
--- /data/deadwood/source/Mesa-7.11-staging/src/gallium/drivers/softpipe/sp_query.c	2010-10-05 18:56:51.000000000 +0200
+++ ./mesa/src/gallium/drivers/softpipe/sp_query.c	2013-03-13 22:19:46.000000000 +0100
@@ -157,7 +157,7 @@
       /*os_get_time is in microseconds*/
       td.frequency = 1000000;
       td.disjoint = FALSE;
-      memcpy(vresult, &sq->so,
+      memcpy(vresult, &td,
              sizeof(struct pipe_query_data_timestamp_disjoint));
    }
       break;
diff -ur '--exclude-from=./mesa/src/aros/mesa.diff.excluded' -I '^.*\bRevision.*Date.*\b.*$' /data/deadwood/source/Mesa-7.11-staging/src/gallium/include/pipe/p_config.h ./mesa/src/gallium/include/pipe/p_config.h
--- /data/deadwood/source/Mesa-7.11-staging/src/gallium/include/pipe/p_config.h	2011-08-01 05:23:42.000000000 +0200
+++ ./mesa/src/gallium/include/pipe/p_config.h	2013-03-13 22:19:46.000000000 +0100
@@ -82,10 +82,18 @@
 #define PIPE_ARCH_X86
 #endif
 
+#if defined(__arm__)
+#define PIPE_ARCH_ARM
+#endif
+
 #if defined(__x86_64__) /* gcc */ || defined(_M_X64) /* msvc */ || defined(_M_AMD64) /* msvc */ || defined(__x86_64) /* Sun cc */
 #define PIPE_ARCH_X86_64
 #endif
 
+#if defined(__mc68000) /* gcc */
+#define PIPE_ARCH_M68K
+#endif
+
 #if defined(PIPE_ARCH_X86) || defined(PIPE_ARCH_X86_64)
 #if defined(PIPE_CC_GCC) && !defined(__SSE2__)
 /* #warning SSE2 support requires -msse -msse2 compiler options */
@@ -99,7 +107,7 @@
 #endif
 #endif
 
-#if defined(__ppc__) || defined(__ppc64__) || defined(__PPC__)
+#if defined(__ppc__) || defined(__ppc64__) || defined(__PPC__) || defined(__powerpc__)
 #define PIPE_ARCH_PPC
 #if defined(__ppc64__) || defined(__PPC64__)
 #define PIPE_ARCH_PPC_64
@@ -131,9 +139,9 @@
 
 #else
 
-#if defined(PIPE_ARCH_X86) || defined(PIPE_ARCH_X86_64)
+#if defined(PIPE_ARCH_X86) || defined(PIPE_ARCH_X86_64) || defined(PIPE_ARCH_ARM)
 #define PIPE_ARCH_LITTLE_ENDIAN
-#elif defined(PIPE_ARCH_PPC) || defined(PIPE_ARCH_PPC_64)
+#elif defined(PIPE_ARCH_PPC) || defined(PIPE_ARCH_PPC_64) || defined(PIPE_ARCH_M68K)
 #define PIPE_ARCH_BIG_ENDIAN
 #endif
 
@@ -149,6 +157,15 @@
  * See subsystem below for a more fine-grained distinction.
  */
 
+/*
+ * In some situations __AROS__ definition can coexist with anything of
+ * the below (for example this happens on ARM port which is compiled using Linux
+ * compiler at the moment). So if we have __AROS__ we don't evaluate anything else
+ */
+#if defined(__AROS__)
+#define PIPE_OS_AROS
+#else
+
 #if defined(__linux__)
 #define PIPE_OS_LINUX
 #define PIPE_OS_UNIX
@@ -201,6 +218,8 @@
 #define PIPE_OS_UNIX
 #endif
 
+#endif
+
 /*
  * Try to auto-detect the subsystem.
  * 
diff -ur '--exclude-from=./mesa/src/aros/mesa.diff.excluded' -I '^.*\bRevision.*Date.*\b.*$' /data/deadwood/source/Mesa-7.11-staging/src/gallium/state_trackers/egl/common/egl_g3d_api.c ./mesa/src/gallium/state_trackers/egl/common/egl_g3d_api.c
--- /data/deadwood/source/Mesa-7.11-staging/src/gallium/state_trackers/egl/common/egl_g3d_api.c	2011-07-09 03:37:09.000000000 +0200
+++ ./mesa/src/gallium/state_trackers/egl/common/egl_g3d_api.c	2012-01-22 09:28:44.000000000 +0100
@@ -86,7 +86,7 @@
    return stapi;
 }
 
-static int
+static EGLint
 egl_g3d_compare_config(const _EGLConfig *conf1, const _EGLConfig *conf2,
                        void *priv_data)
 {
@@ -125,6 +125,28 @@
 
    if (!num_configs)
       return _eglError(EGL_BAD_PARAMETER, "eglChooseConfigs");
+#if defined(_EGL_OS_AROS)
+    /* This is the earliest place where client indicates which API is it interested in. Due to fact that
+     * HostGL is not providing Gallium st_api, we detected if we are capable of continuing with Gallium EGL
+     * driver and if not we "chain load" the display into AROSMesa EGL driver
+     */
+   int j = -1;
+   while (attribs[++j] != EGL_NONE)
+   {
+       if (attribs[j] == EGL_RENDERABLE_TYPE)
+       {
+           if (attribs[++j] == EGL_OPENGL_BIT)
+           {
+               if (egl_g3d_get_st_api(drv, EGL_OPENGL_BIT) == NULL)
+               {
+                   dpy->Driver = _eglBuiltInDriverAROSMesa(NULL);
+                   dpy->Driver->API.Initialize(dpy->Driver, dpy);
+                   return dpy->Driver->API.ChooseConfig(dpy->Driver, dpy,attribs, configs, size, num_configs);
+               }
+           }
+       }
+   }
+#endif
 
    if (!_eglParseConfigAttribList(&criteria, dpy, attribs))
       return _eglError(EGL_BAD_ATTRIBUTE, "eglChooseConfig");
@@ -247,6 +269,7 @@
    struct egl_g3d_surface *gsurf;
    struct native_surface *nsurf;
    const char *err;
+   int w,h;
 
    switch (arg->type) {
    case EGL_WINDOW_BIT:
@@ -304,8 +327,9 @@
       return NULL;
    }
    /* initialize the geometry */
-   if (!nsurf->validate(nsurf, 0x0, &gsurf->sequence_number, NULL,
-            &gsurf->base.Width, &gsurf->base.Height)) {
+   if (!nsurf->validate(nsurf, 0x0, &gsurf->sequence_number, NULL, &w, &h)) {
+      gsurf->base.Width = w;
+      gsurf->base.Height = h;
       nsurf->destroy(nsurf);
       FREE(gsurf);
       return NULL;
diff -ur '--exclude-from=./mesa/src/aros/mesa.diff.excluded' -I '^.*\bRevision.*Date.*\b.*$' /data/deadwood/source/Mesa-7.11-staging/src/gallium/state_trackers/egl/common/egl_g3d.c ./mesa/src/gallium/state_trackers/egl/common/egl_g3d.c
--- /data/deadwood/source/Mesa-7.11-staging/src/gallium/state_trackers/egl/common/egl_g3d.c	2011-07-09 03:37:09.000000000 +0200
+++ ./mesa/src/gallium/state_trackers/egl/common/egl_g3d.c	2011-09-05 20:03:01.000000000 +0200
@@ -56,6 +56,7 @@
       gctx->stctxi->notify_invalid_framebuffer(gctx->stctxi, gsurf->stfbi);
 }
 
+#if !defined(PIPE_OS_AROS)
 static struct pipe_screen *
 egl_g3d_new_drm_screen(struct native_display *ndpy, const char *name, int fd)
 {
@@ -71,6 +72,7 @@
    struct egl_g3d_display *gdpy = egl_g3d_display(dpy);
    return gdpy->loader->create_sw_screen(ws);
 }
+#endif
 
 static struct pipe_resource *
 egl_g3d_lookup_egl_image(struct native_display *ndpy, void *egl_image)
@@ -87,13 +89,19 @@
    return resource;
 }
 
+#if !defined(PIPE_OS_AROS)
 static const struct native_event_handler egl_g3d_native_event_handler = {
    egl_g3d_invalid_surface,
    egl_g3d_new_drm_screen,
    egl_g3d_new_sw_screen,
    egl_g3d_lookup_egl_image
 };
-
+#else
+static const struct native_event_handler egl_g3d_native_event_handler = {
+   egl_g3d_invalid_surface,
+   egl_g3d_lookup_egl_image
+};
+#endif
 /**
  * Get the native platform.
  */
@@ -137,6 +145,12 @@
          nplat = native_get_fbdev_platform(&egl_g3d_native_event_handler);
 #endif
          break;
+      case _EGL_PLATFORM_AROS:
+         plat_name = "AROS";
+#ifdef HAVE_AROS_BACKEND
+         nplat = native_get_aros_platform(&egl_g3d_native_event_handler);
+#endif
+         break;
       default:
          break;
       }
@@ -157,7 +171,7 @@
 {
    struct egl_g3d_display *gdpy = egl_g3d_display(dpy);
    const struct native_connector **native_connectors;
-   EGLint num_connectors, i;
+   int num_connectors, i;
 
    native_connectors =
       gdpy->native->modeset->get_connectors(gdpy->native, &num_connectors, NULL);
@@ -171,7 +185,7 @@
       const struct native_connector *nconn = native_connectors[i];
       struct egl_g3d_screen *gscr;
       const struct native_mode **native_modes;
-      EGLint num_modes, j;
+      int num_modes, j;
 
       /* TODO support for hotplug */
       native_modes =
diff -ur '--exclude-from=./mesa/src/aros/mesa.diff.excluded' -I '^.*\bRevision.*Date.*\b.*$' /data/deadwood/source/Mesa-7.11-staging/src/gallium/state_trackers/egl/common/egl_g3d_loader.h ./mesa/src/gallium/state_trackers/egl/common/egl_g3d_loader.h
--- /data/deadwood/source/Mesa-7.11-staging/src/gallium/state_trackers/egl/common/egl_g3d_loader.h	2011-07-09 03:37:09.000000000 +0200
+++ ./mesa/src/gallium/state_trackers/egl/common/egl_g3d_loader.h	2011-09-05 20:03:01.000000000 +0200
@@ -40,8 +40,10 @@
    uint profile_masks[ST_API_COUNT];
    struct st_api *(*get_st_api)(enum st_api_type api);
 
+#if !defined(PIPE_OS_AROS)
    struct pipe_screen *(*create_drm_screen)(const char *name, int fd);
    struct pipe_screen *(*create_sw_screen)(struct sw_winsys *ws);
+#endif
 };
 
 _EGLDriver *
diff -ur '--exclude-from=./mesa/src/aros/mesa.diff.excluded' -I '^.*\bRevision.*Date.*\b.*$' /data/deadwood/source/Mesa-7.11-staging/src/gallium/state_trackers/egl/common/egl_g3d_st.c ./mesa/src/gallium/state_trackers/egl/common/egl_g3d_st.c
--- /data/deadwood/source/Mesa-7.11-staging/src/gallium/state_trackers/egl/common/egl_g3d_st.c	2011-07-07 02:04:59.000000000 +0200
+++ ./mesa/src/gallium/state_trackers/egl/common/egl_g3d_st.c	2011-09-05 20:03:01.000000000 +0200
@@ -207,6 +207,7 @@
    struct pipe_resource *textures[NUM_NATIVE_ATTACHMENTS];
    uint attachment_mask = 0;
    unsigned i;
+   int w, h;
 
    for (i = 0; i < count; i++) {
       int natt;
@@ -234,10 +235,12 @@
    }
 
    if (!gsurf->native->validate(gsurf->native, attachment_mask,
-         &gsurf->sequence_number, textures, &gsurf->base.Width,
-         &gsurf->base.Height))
+         &gsurf->sequence_number, textures, &w, &h)) {
+      gsurf->base.Width = w;
+      gsurf->base.Height = h;
       return FALSE;
-
+   }
+   
    for (i = 0; i < count; i++) {
       struct pipe_resource *tex;
       int natt;
diff -ur '--exclude-from=./mesa/src/aros/mesa.diff.excluded' -I '^.*\bRevision.*Date.*\b.*$' /data/deadwood/source/Mesa-7.11-staging/src/gallium/state_trackers/egl/common/native.h ./mesa/src/gallium/state_trackers/egl/common/native.h
--- /data/deadwood/source/Mesa-7.11-staging/src/gallium/state_trackers/egl/common/native.h	2011-07-09 03:37:09.000000000 +0200
+++ ./mesa/src/gallium/state_trackers/egl/common/native.h	2011-09-05 20:03:01.000000000 +0200
@@ -220,10 +220,12 @@
                            struct native_surface *nsurf,
                            unsigned int seq_num);
 
+#if !defined(PIPE_OS_AROS)
    struct pipe_screen *(*new_drm_screen)(struct native_display *ndpy,
                                          const char *name, int fd);
    struct pipe_screen *(*new_sw_screen)(struct native_display *ndpy,
                                         struct sw_winsys *ws);
+#endif
 
    struct pipe_resource *(*lookup_egl_image)(struct native_display *ndpy,
                                              void *egl_image);
@@ -288,8 +290,14 @@
 const struct native_platform *
 native_get_fbdev_platform(const struct native_event_handler *event_handler);
 
+#if defined(PIPE_OS_AROS)
+const struct native_platform *
+native_get_aros_platform(const struct native_event_handler *event_handler);
+#endif
+
 #ifdef __cplusplus
 }
 #endif
 
+
 #endif /* _NATIVE_H_ */
diff -ur '--exclude-from=./mesa/src/aros/mesa.diff.excluded' -I '^.*\bRevision.*Date.*\b.*$' /data/deadwood/source/Mesa-7.11-staging/src/gallium/state_trackers/egl/common/native_helper.c ./mesa/src/gallium/state_trackers/egl/common/native_helper.c
--- /data/deadwood/source/Mesa-7.11-staging/src/gallium/state_trackers/egl/common/native_helper.c	2011-07-09 03:37:09.000000000 +0200
+++ ./mesa/src/gallium/state_trackers/egl/common/native_helper.c	2013-03-13 22:19:46.000000000 +0100
@@ -282,10 +282,10 @@
 			      btex, 0, &src_box);
    ret = TRUE;
 
- out_no_ftex:
    pipe_resource_reference(&btex, NULL);
  out_no_btex:
    pipe_resource_reference(&ftex, NULL);
+ out_no_ftex:
 
    return ret;
 }
diff -ur '--exclude-from=./mesa/src/aros/mesa.diff.excluded' -I '^.*\bRevision.*Date.*\b.*$' /data/deadwood/source/Mesa-7.11-staging/src/gallium/state_trackers/vega/mask.c ./mesa/src/gallium/state_trackers/vega/mask.c
--- /data/deadwood/source/Mesa-7.11-staging/src/gallium/state_trackers/vega/mask.c	2011-07-09 03:37:09.000000000 +0200
+++ ./mesa/src/gallium/state_trackers/vega/mask.c	2011-09-05 20:03:01.000000000 +0200
@@ -384,9 +384,11 @@
                      VGint width, VGint height,
                      VGfloat value)
 {
+#if 0 /* unused */
    VGfloat alpha_color[4] = {0, 0, 0, 0};
 
    alpha_color[3] = value;
+#endif
 
    mask_resource_fill(layer->sampler_view->texture,
                       x, y, width, height, value);
diff -ur '--exclude-from=./mesa/src/aros/mesa.diff.excluded' -I '^.*\bRevision.*Date.*\b.*$' /data/deadwood/source/Mesa-7.11-staging/src/gallium/state_trackers/vega/path.c ./mesa/src/gallium/state_trackers/vega/path.c
--- /data/deadwood/source/Mesa-7.11-staging/src/gallium/state_trackers/vega/path.c	2011-01-23 03:05:41.000000000 +0100
+++ ./mesa/src/gallium/state_trackers/vega/path.c	2011-09-05 20:03:01.000000000 +0200
@@ -1093,8 +1093,8 @@
       y2 = data[1];
       x3 = data[2];
       y3 = data[3];
-      map_if_relative(pd->ox, pd->oy, relative, &x2, &y2);
-      map_if_relative(pd->ox, pd->oy, relative, &x3, &y3);
+      map_if_relative(x0, y0, relative, &x2, &y2);
+      map_if_relative(x0, y0, relative, &x3, &y3);
       pd->ox = x3;
       pd->oy = y3;
       pd->px = x2;
diff -ur '--exclude-from=./mesa/src/aros/mesa.diff.excluded' -I '^.*\bRevision.*Date.*\b.*$' /data/deadwood/source/Mesa-7.11-staging/src/gallium/targets/egl-static/egl.c ./mesa/src/gallium/targets/egl-static/egl.c
--- /data/deadwood/source/Mesa-7.11-staging/src/gallium/targets/egl-static/egl.c	2011-07-09 03:37:09.000000000 +0200
+++ ./mesa/src/gallium/targets/egl-static/egl.c	2011-09-05 20:03:01.000000000 +0200
@@ -60,6 +60,7 @@
    return stmod->stapi;
 }
 
+#if !defined(PIPE_OS_AROS)
 static const char *
 drm_fd_get_screen_name(int fd)
 {
@@ -145,6 +146,7 @@
 {
    return egl_pipe_create_swrast_screen(ws);
 }
+#endif
 
 static const struct egl_g3d_loader *
 loader_init(void)
@@ -155,8 +157,10 @@
       egl_g3d_loader.profile_masks[i] = egl_st_get_profile_mask(i);
 
    egl_g3d_loader.get_st_api = get_st_api;
+#if !defined(PIPE_OS_AROS)
    egl_g3d_loader.create_drm_screen = create_drm_screen;
    egl_g3d_loader.create_sw_screen = create_sw_screen;
+#endif
 
    return &egl_g3d_loader;
 }
diff -ur '--exclude-from=./mesa/src/aros/mesa.diff.excluded' -I '^.*\bRevision.*Date.*\b.*$' /data/deadwood/source/Mesa-7.11-staging/src/gallium/targets/egl-static/st_GL.c ./mesa/src/gallium/targets/egl-static/st_GL.c
--- /data/deadwood/source/Mesa-7.11-staging/src/gallium/targets/egl-static/st_GL.c	2011-07-09 03:37:09.000000000 +0200
+++ ./mesa/src/gallium/targets/egl-static/st_GL.c	2012-01-21 20:54:03.000000000 +0100
@@ -1,38 +1,33 @@
-/*
- * Mesa 3-D graphics library
- * Version:  7.10
- *
- * Copyright (C) 2011 LunarG Inc.
- *
- * Permission is hereby granted, free of charge, to any person obtaining a
- * copy of this software and associated documentation files (the "Software"),
- * to deal in the Software without restriction, including without limitation
- * the rights to use, copy, modify, merge, publish, distribute, sublicense,
- * and/or sell copies of the Software, and to permit persons to whom the
- * Software is furnished to do so, subject to the following conditions:
- *
- * The above copyright notice and this permission notice shall be included
- * in all copies or substantial portions of the Software.
- *
- * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
- * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
- * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL
- * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
- * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
- * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
- * DEALINGS IN THE SOFTWARE.
- *
- * Authors:
- *    Chia-I Wu <olv@lunarg.com>
- */
-#include "state_tracker/st_gl_api.h"
-#include "pipe/p_compiler.h"
+#include <proto/exec.h>
+#if !defined(__mc68000)
+#include <proto/mesa.h>
+#endif
+#include <aros/symbolsets.h>
 
-PUBLIC struct st_api *
-st_api_create_OpenGL(void);
+#include "egl_st.h"
 
-struct st_api *
-st_api_create_OpenGL(void)
+struct Library * MesaBase = NULL;
+
+PUBLIC struct st_api * st_gl_api_create(void)
+{
+#if !defined(__mc68000)
+    if (!MesaBase)
+        MesaBase = OpenLibrary("mesa.library", 0L);
+
+    if (MesaBase)
+        return (struct st_api *) AROSMesaGetOpenGLStateTrackerApi();
+    else
+#endif
+        return NULL;
+}
+
+static VOID CloseMesa()
 {
-   return st_gl_api_create();
+    if (MesaBase)
+    {
+        CloseLibrary(MesaBase);
+        MesaBase = NULL;
+    }
 }
+
+ADD2EXPUNGELIB(CloseMesa, 5)
diff -ur '--exclude-from=./mesa/src/aros/mesa.diff.excluded' -I '^.*\bRevision.*Date.*\b.*$' /data/deadwood/source/Mesa-7.11-staging/src/glsl/ir_function.cpp ./mesa/src/glsl/ir_function.cpp
--- /data/deadwood/source/Mesa-7.11-staging/src/glsl/ir_function.cpp	2011-07-20 01:39:56.000000000 +0200
+++ ./mesa/src/glsl/ir_function.cpp	2013-01-04 19:25:40.000000000 +0100
@@ -111,7 +111,7 @@
        * but the actual parameter can be coerced to the type of the declared
        * parameter, the match score is one.
        */
-      int score;
+      int score = -1;
       switch ((enum ir_variable_mode)(param->mode)) {
       case ir_var_auto:
       case ir_var_uniform:
diff -ur '--exclude-from=./mesa/src/aros/mesa.diff.excluded' -I '^.*\bRevision.*Date.*\b.*$' /data/deadwood/source/Mesa-7.11-staging/src/glsl/ir_function_detect_recursion.cpp ./mesa/src/glsl/ir_function_detect_recursion.cpp
--- /data/deadwood/source/Mesa-7.11-staging/src/glsl/ir_function_detect_recursion.cpp	2011-07-23 10:54:07.000000000 +0200
+++ ./mesa/src/glsl/ir_function_detect_recursion.cpp	2013-10-03 18:47:38.000000000 +0200
@@ -238,7 +238,11 @@
    bool progress;
 };
 
-static void
+/* AROS: Removing static fixes an invalid compilation under
+ * gcc 4.6.2 -O2 mode where code generated remove_unlinked_functions
+ * in would loop indefinatelly
+ */
+void
 destroy_links(exec_list *list, function *f)
 {
    foreach_list_safe(node, list) {
diff -ur '--exclude-from=./mesa/src/aros/mesa.diff.excluded' -I '^.*\bRevision.*Date.*\b.*$' /data/deadwood/source/Mesa-7.11-staging/src/glsl/ir.h ./mesa/src/glsl/ir.h
--- /data/deadwood/source/Mesa-7.11-staging/src/glsl/ir.h	2011-07-23 10:54:07.000000000 +0200
+++ ./mesa/src/glsl/ir.h	2013-01-04 19:25:40.000000000 +0100
@@ -41,6 +41,14 @@
  * @{
  */
 
+#if defined(__AROS__)
+#include <aros/debug.h>
+#undef VOLATILE
+#undef STATIC
+#define log2(x)             ((float) (log(x) * 1.442695f))
+#define printf(fmt, ...)    bug(fmt, ##__VA_ARGS__)
+#endif
+
 /**
  * Class tags
  *
diff -ur '--exclude-from=./mesa/src/aros/mesa.diff.excluded' -I '^.*\bRevision.*Date.*\b.*$' /data/deadwood/source/Mesa-7.11-staging/src/glsl/ralloc.c ./mesa/src/glsl/ralloc.c
--- /data/deadwood/source/Mesa-7.11-staging/src/glsl/ralloc.c	2011-07-09 03:31:41.000000000 +0200
+++ ./mesa/src/glsl/ralloc.c	2011-09-05 20:03:01.000000000 +0200
@@ -28,6 +28,10 @@
 #include <string.h>
 #include <stdint.h>
 
+#if defined(__AROS__)
+#define SIZE_MAX INT32_MAX
+#endif
+
 #include "ralloc.h"
 
 #ifdef __GNUC__
diff -ur '--exclude-from=./mesa/src/aros/mesa.diff.excluded' -I '^.*\bRevision.*Date.*\b.*$' /data/deadwood/source/Mesa-7.11-staging/src/glsl/ralloc.h ./mesa/src/glsl/ralloc.h
--- /data/deadwood/source/Mesa-7.11-staging/src/glsl/ralloc.h	2011-07-09 03:31:41.000000000 +0200
+++ ./mesa/src/glsl/ralloc.h	2011-09-05 20:03:01.000000000 +0200
@@ -53,7 +53,9 @@
 
 #include <stddef.h>
 #include <stdarg.h>
+#ifndef __cplusplus
 #include <stdbool.h>
+#endif
 
 /**
  * \def ralloc(ctx, type)
diff -ur '--exclude-from=./mesa/src/aros/mesa.diff.excluded' -I '^.*\bRevision.*Date.*\b.*$' /data/deadwood/source/Mesa-7.11-staging/src/glsl/s_expression.cpp ./mesa/src/glsl/s_expression.cpp
--- /data/deadwood/source/Mesa-7.11-staging/src/glsl/s_expression.cpp	2011-07-09 03:32:30.000000000 +0200
+++ ./mesa/src/glsl/s_expression.cpp	2011-09-05 20:03:01.000000000 +0200
@@ -25,6 +25,11 @@
 #include <assert.h>
 #include "s_expression.h"
 
+#if defined(__AROS__)
+#include <aros/debug.h>
+#define printf(fmt, ...)    bug(fmt, ##__VA_ARGS__)
+#endif
+
 s_symbol::s_symbol(const char *tmp, size_t n)
 {
    this->str = ralloc_strndup (this, tmp, n);
diff -ur '--exclude-from=./mesa/src/aros/mesa.diff.excluded' -I '^.*\bRevision.*Date.*\b.*$' /data/deadwood/source/Mesa-7.11-staging/src/glu/sgi/include/gluos.h ./mesa/src/glu/sgi/include/gluos.h
--- /data/deadwood/source/Mesa-7.11-staging/src/glu/sgi/include/gluos.h	2011-07-09 03:37:10.000000000 +0200
+++ ./mesa/src/glu/sgi/include/gluos.h	2011-09-05 20:03:01.000000000 +0200
@@ -2,6 +2,13 @@
 ** gluos.h - operating system dependencies for GLU
 **
 */
+
+#if defined(__AROS__)
+#include <GL/glu.h>
+#include "glu/gluapim.h"
+#include "glu/glu_mangle.h"
+#endif
+
 #ifdef __VMS
 #ifdef __cplusplus
 #pragma message disable nocordel
diff -ur '--exclude-from=./mesa/src/aros/mesa.diff.excluded' -I '^.*\bRevision.*Date.*\b.*$' /data/deadwood/source/Mesa-7.11-staging/src/glu/sgi/libutil/mipmap.c ./mesa/src/glu/sgi/libutil/mipmap.c
--- /data/deadwood/source/Mesa-7.11-staging/src/glu/sgi/libutil/mipmap.c	2011-07-09 03:37:10.000000000 +0200
+++ ./mesa/src/glu/sgi/libutil/mipmap.c	2011-09-05 20:03:01.000000000 +0200
@@ -4886,7 +4886,7 @@
     GLint group_size;
     GLint elements_per_line;
     const GLubyte *start;
-    const GLubyte *iter;
+    const GLubyte *iter = NULL;
     GLushort *iter2;
     GLint i, j, k;
     GLint myswap_bytes;
@@ -5149,7 +5149,7 @@
     GLint group_size;
     GLint elements_per_line;
     GLubyte *start;
-    GLubyte *iter;
+    GLubyte *iter = NULL;
     const GLushort *iter2;
     GLint i, j, k;
     GLint myswap_bytes;
@@ -6034,7 +6034,7 @@
     int i,j,k,xindex;
 
     const char *temp, *temp0;
-    int outindex;
+    int outindex = 0;
 
     int lowx_int, highx_int, lowy_int, highy_int;
     float x_percent, y_percent;
@@ -6695,7 +6695,7 @@
    int elementsPerLine;
    int rowsPerImage;
    int imageSize;
-   const GLubyte *start, *rowStart, *iter;
+   const GLubyte *start, *rowStart, *iter = NULL;
    GLushort *iter2;
    int ww, hh, dd, k;
 
@@ -7057,7 +7057,7 @@
    int groupSize;
    int rowSize;
    int padding;
-   GLubyte *start, *rowStart, *iter;
+   GLubyte *start, *rowStart, *iter = NULL;
    int elementsPerLine;
    const GLushort *iter2;
    int ii, jj, dd, k;
diff -ur '--exclude-from=./mesa/src/aros/mesa.diff.excluded' -I '^.*\bRevision.*Date.*\b.*$' /data/deadwood/source/Mesa-7.11-staging/src/mapi/glapi/gen/gl_API.xml ./mesa/src/mapi/glapi/gen/gl_API.xml
--- /data/deadwood/source/Mesa-7.11-staging/src/mapi/glapi/gen/gl_API.xml	2011-07-09 03:37:10.000000000 +0200
+++ ./mesa/src/mapi/glapi/gen/gl_API.xml	2011-09-05 20:03:01.000000000 +0200
@@ -12318,7 +12318,7 @@
     <function name="TextureRangeAPPLE" offset="assign" static_dispatch="false">
         <param name="target" type="GLenum"/>
         <param name="length" type="GLsizei"/>
-        <param name="pointer" type="GLvoid *"/>
+        <param name="pointer" type="const GLvoid *"/>
     </function>
     <function name="GetTexParameterPointervAPPLE" offset="assign" static_dispatch="false">
         <param name="target" type="GLenum"/>
diff -ur '--exclude-from=./mesa/src/aros/mesa.diff.excluded' -I '^.*\bRevision.*Date.*\b.*$' /data/deadwood/source/Mesa-7.11-staging/src/mapi/glapi/gen/gl_procs.py ./mesa/src/mapi/glapi/gen/gl_procs.py
--- /data/deadwood/source/Mesa-7.11-staging/src/mapi/glapi/gen/gl_procs.py	2011-01-06 01:19:15.000000000 +0100
+++ ./mesa/src/mapi/glapi/gen/gl_procs.py	2011-09-08 18:20:04.000000000 +0200
@@ -107,7 +107,11 @@
 		for func in api.functionIterateByOffset():
 			for n in func.entry_points:
 				if n != func.name:
-					name = func.dispatch_name()
+					# AROS: call into a static dispach instead of alliased function dispatch
+					# This will be needed as long as gl_procs.h is used to generate the
+					# AROSMesaGetProcAddress function table which is located in libGL.a
+					# name = func.dispatch_name()
+					name = n
 					self.printFunctionString( n )
 					
 					if func.has_different_protocol(n):
diff -ur '--exclude-from=./mesa/src/aros/mesa.diff.excluded' -I '^.*\bRevision.*Date.*\b.*$' /data/deadwood/source/Mesa-7.11-staging/src/mapi/glapi/gen/gl_XML.py ./mesa/src/mapi/glapi/gen/gl_XML.py
--- /data/deadwood/source/Mesa-7.11-staging/src/mapi/glapi/gen/gl_XML.py	2011-07-09 03:37:10.000000000 +0200
+++ ./mesa/src/mapi/glapi/gen/gl_XML.py	2011-09-05 20:03:01.000000000 +0200
@@ -629,8 +629,9 @@
 		name = element.nsProp( "name", None )
 		alias = element.nsProp( "alias", None )
 
-		if is_attr_true(element, "static_dispatch"):
-			self.static_entry_points.append(name)
+		# if is_attr_true(element, "static_dispatch"):
+		# AROS version requires all entry points to be available in library base
+		self.static_entry_points.append( name )
 
 		self.entry_points.append( name )
 		if alias:
diff -ur '--exclude-from=./mesa/src/aros/mesa.diff.excluded' -I '^.*\bRevision.*Date.*\b.*$' /data/deadwood/source/Mesa-7.11-staging/src/mapi/glapi/gen/Makefile ./mesa/src/mapi/glapi/gen/Makefile
--- /data/deadwood/source/Mesa-7.11-staging/src/mapi/glapi/gen/Makefile	2011-07-09 03:37:10.000000000 +0200
+++ ./mesa/src/mapi/glapi/gen/Makefile	2011-09-05 20:03:01.000000000 +0200
@@ -5,7 +5,9 @@
 
 
 TOP = ../../../..
-include $(TOP)/configs/current
+#include $(TOP)/configs/current
+PYTHON2 = python
+PYTHON_FLAGS = -t -O -O
 
 MESA_DIR = $(TOP)/src/mesa
 MESA_GLAPI_DIR = $(TOP)/src/mapi/glapi
@@ -30,11 +32,12 @@
 	$(MESA_DIR)/main/enums.c \
 	$(MESA_DIR)/main/dispatch.h \
 	$(MESA_DIR)/main/remap_helper.h \
-	$(MESA_GLX_DIR)/indirect.c \
-	$(MESA_GLX_DIR)/indirect.h \
-	$(MESA_GLX_DIR)/indirect_init.c \
-	$(MESA_GLX_DIR)/indirect_size.h \
-	$(MESA_GLX_DIR)/indirect_size.c
+
+#	$(MESA_GLX_DIR)/indirect.c \
+#	$(MESA_GLX_DIR)/indirect.h \
+#	$(MESA_GLX_DIR)/indirect_init.c \
+#	$(MESA_GLX_DIR)/indirect_size.h \
+#	$(MESA_GLX_DIR)/indirect_size.c
 
 ######################################################################
 
diff -ur '--exclude-from=./mesa/src/aros/mesa.diff.excluded' -I '^.*\bRevision.*Date.*\b.*$' /data/deadwood/source/Mesa-7.11-staging/src/mapi/glapi/glapi_entrypoint.c ./mesa/src/mapi/glapi/glapi_entrypoint.c
--- /data/deadwood/source/Mesa-7.11-staging/src/mapi/glapi/glapi_entrypoint.c	2010-10-19 19:58:29.000000000 +0200
+++ ./mesa/src/mapi/glapi/glapi_entrypoint.c	2011-09-05 20:03:01.000000000 +0200
@@ -309,10 +309,12 @@
 
 #else /* USE_*_ASM */
 
+#if defined(PTHREADS) || defined(GLX_USE_TLS)
 static void
 init_glapi_relocs( void )
 {
 }
+#endif
 
 
 _glapi_proc
diff -ur '--exclude-from=./mesa/src/aros/mesa.diff.excluded' -I '^.*\bRevision.*Date.*\b.*$' /data/deadwood/source/Mesa-7.11-staging/src/mapi/glapi/glapi_gentable.c ./mesa/src/mapi/glapi/glapi_gentable.c
--- /data/deadwood/source/Mesa-7.11-staging/src/mapi/glapi/glapi_gentable.c	2011-07-09 03:37:10.000000000 +0200
+++ ./mesa/src/mapi/glapi/glapi_gentable.c	2011-09-07 20:43:14.000000000 +0200
@@ -27,6 +27,7 @@
  * SOFTWARE.
  */
 
+#if !defined(__AROS__)
 
 #if defined(DEBUG) && !defined(_WIN32_WCE)
 #include <execinfo.h>
@@ -8727,6 +8728,13 @@
 
     if(!disp->FramebufferTextureLayerEXT) {
         void ** procp = (void **) &disp->FramebufferTextureLayerEXT;
+        snprintf(symboln, sizeof(symboln), "%sFramebufferTextureLayerARB", symbol_prefix);
+        *procp = dlsym(handle, symboln);
+    }
+
+
+    if(!disp->FramebufferTextureLayerEXT) {
+        void ** procp = (void **) &disp->FramebufferTextureLayerEXT;
         snprintf(symboln, sizeof(symboln), "%sFramebufferTextureLayerEXT", symbol_prefix);
         *procp = dlsym(handle, symboln);
     }
@@ -9135,4 +9143,5 @@
 
     return disp;
 }
+#endif
 
diff -ur '--exclude-from=./mesa/src/aros/mesa.diff.excluded' -I '^.*\bRevision.*Date.*\b.*$' /data/deadwood/source/Mesa-7.11-staging/src/mapi/glapi/glapi.h ./mesa/src/mapi/glapi/glapi.h
--- /data/deadwood/source/Mesa-7.11-staging/src/mapi/glapi/glapi.h	2011-07-09 03:37:10.000000000 +0200
+++ ./mesa/src/mapi/glapi/glapi.h	2011-09-05 20:03:01.000000000 +0200
@@ -92,6 +92,17 @@
 # define GET_DISPATCH() _glapi_tls_Dispatch
 # define GET_CURRENT_CONTEXT(C)  struct gl_context *C = (struct gl_context *) _glapi_tls_Context
 
+#elif defined(__AROS__)
+
+#if !defined(AROSMESA_NO_MANGLING)
+#include <GL/gl.h>
+#include "arosmesa/arosmesaapim.h"
+#include "arosmesa/arosmesa_mangle.h"
+#endif
+
+#define GET_DISPATCH() _glapi_get_dispatch()
+#define GET_CURRENT_CONTEXT(C)  struct gl_context *C = (struct gl_context *) _glapi_get_context()
+
 #else
 
 _GLAPI_EXPORT extern struct _glapi_table *_glapi_Dispatch;
diff -ur '--exclude-from=./mesa/src/aros/mesa.diff.excluded' -I '^.*\bRevision.*Date.*\b.*$' /data/deadwood/source/Mesa-7.11-staging/src/mapi/mapi/u_current.c ./mesa/src/mapi/mapi/u_current.c
--- /data/deadwood/source/Mesa-7.11-staging/src/mapi/mapi/u_current.c	2011-07-09 03:37:10.000000000 +0200
+++ ./mesa/src/mapi/mapi/u_current.c	2011-12-17 11:25:59.000000000 +0100
@@ -108,9 +108,17 @@
 
 #else
 
+#if defined(__AROS__)
+#include "aros/tls.h"
+
+DECLARE_STATIC_TLS(u_current_table);
+DECLARE_STATIC_TLS(u_current_user);
+
+#else
 struct mapi_table *u_current_table =
    (struct mapi_table *) table_noop_array;
 void *u_current_user;
+#endif
 
 #ifdef THREADS
 struct u_tsd u_current_table_tsd;
@@ -200,6 +208,8 @@
 #elif defined(THREADS)
    u_tsd_set(&u_current_user_tsd, (void *) ptr);
    u_current_user = (ThreadSafe) ? NULL : (void *) ptr;
+#elif defined(__AROS__)
+   InsertIntoTLS(u_current_user, (APTR)ptr);
 #else
    u_current_user = (void *) ptr;
 #endif
@@ -219,6 +229,8 @@
    return (ThreadSafe)
       ? u_tsd_get(&u_current_user_tsd)
       : u_current_user;
+#elif defined(__AROS__)
+   return GetFromTLS(u_current_user);
 #else
    return u_current_user;
 #endif
@@ -244,6 +256,8 @@
 #elif defined(THREADS)
    u_tsd_set(&u_current_table_tsd, (void *) tbl);
    u_current_table = (ThreadSafe) ? NULL : (void *) tbl;
+#elif defined(__AROS__)
+   InsertIntoTLS(u_current_table, (APTR)tbl);
 #else
    u_current_table = (struct mapi_table *) tbl;
 #endif
@@ -260,6 +274,12 @@
 #elif defined(THREADS)
    return (struct mapi_table *) ((ThreadSafe) ?
          u_tsd_get(&u_current_table_tsd) : (void *) u_current_table);
+#elif defined(__AROS__)
+   struct mapi_table *tbl;
+   tbl = (struct mapi_table *)GetFromTLS(u_current_table);
+   if (tbl == NULL)
+       tbl = (struct mapi_table *) table_noop_array;
+   return tbl;
 #else
    return u_current_table;
 #endif
diff -ur '--exclude-from=./mesa/src/aros/mesa.diff.excluded' -I '^.*\bRevision.*Date.*\b.*$' /data/deadwood/source/Mesa-7.11-staging/src/mapi/mapi/u_current.h ./mesa/src/mapi/mapi/u_current.h
--- /data/deadwood/source/Mesa-7.11-staging/src/mapi/mapi/u_current.h	2011-05-13 01:47:18.000000000 +0200
+++ ./mesa/src/mapi/mapi/u_current.h	2011-09-05 20:03:01.000000000 +0200
@@ -36,6 +36,8 @@
 extern __thread void *u_current_user
     __attribute__((tls_model("initial-exec")));
 
+#elif defined(__AROS__)
+
 #else /* GLX_USE_TLS */
 
 extern struct mapi_table *u_current_table;
@@ -68,6 +70,8 @@
 {
 #ifdef GLX_USE_TLS
    return u_current_table;
+#elif defined(__AROS__)
+   return u_current_get_internal();
 #else
    return (likely(u_current_table) ?
          u_current_table : u_current_get_internal());
@@ -79,6 +83,8 @@
 {
 #ifdef GLX_USE_TLS
    return u_current_user;
+#elif defined(__AROS__)
+   return u_current_get_user_internal();
 #else
    return likely(u_current_user) ? u_current_user : u_current_get_user_internal();
 #endif
diff -ur '--exclude-from=./mesa/src/aros/mesa.diff.excluded' -I '^.*\bRevision.*Date.*\b.*$' /data/deadwood/source/Mesa-7.11-staging/src/mesa/main/compiler.h ./mesa/src/mesa/main/compiler.h
--- /data/deadwood/source/Mesa-7.11-staging/src/mesa/main/compiler.h	2011-07-09 03:37:10.000000000 +0200
+++ ./mesa/src/mesa/main/compiler.h	2011-09-05 20:03:01.000000000 +0200
@@ -255,6 +255,8 @@
 #elif defined(__APPLE__)
 #include <CoreFoundation/CFByteOrder.h>
 #define CPU_TO_LE32( x )	CFSwapInt32HostToLittle( x )
+#elif defined(__AROS__)
+#define CPU_TO_LE32( x )    AROS_BE2LONG( x )
 #elif (defined(_AIX) || defined(__blrts))
 static INLINE GLuint CPU_TO_LE32(GLuint x)
 {
@@ -322,6 +324,9 @@
 /**
  * ASSERT macro
  */
+#if defined(__AROS__)
+#undef ASSERT
+#endif
 #if !defined(_WIN32_WCE)
 #if defined(BUILD_FOR_SNAP) && defined(CHECKED)
 #  define ASSERT(X)   _CHECK(X) 
diff -ur '--exclude-from=./mesa/src/aros/mesa.diff.excluded' -I '^.*\bRevision.*Date.*\b.*$' /data/deadwood/source/Mesa-7.11-staging/src/mesa/main/dispatch.h ./mesa/src/mesa/main/dispatch.h
--- /data/deadwood/source/Mesa-7.11-staging/src/mesa/main/dispatch.h	2011-07-09 03:37:10.000000000 +0200
+++ ./mesa/src/mesa/main/dispatch.h	2011-09-07 20:43:00.000000000 +0200
@@ -12099,14 +12099,14 @@
    SET_by_offset(disp, _gloffset_GetTexParameterPointervAPPLE, fn);
 }
 
-typedef void (GLAPIENTRYP _glptr_TextureRangeAPPLE)(GLenum, GLsizei, GLvoid *);
+typedef void (GLAPIENTRYP _glptr_TextureRangeAPPLE)(GLenum, GLsizei, const GLvoid *);
 #define CALL_TextureRangeAPPLE(disp, parameters) \
     (* GET_TextureRangeAPPLE(disp)) parameters
 static INLINE _glptr_TextureRangeAPPLE GET_TextureRangeAPPLE(struct _glapi_table *disp) {
    return (_glptr_TextureRangeAPPLE) (GET_by_offset(disp, _gloffset_TextureRangeAPPLE));
 }
 
-static INLINE void SET_TextureRangeAPPLE(struct _glapi_table *disp, void (GLAPIENTRYP fn)(GLenum, GLsizei, GLvoid *)) {
+static INLINE void SET_TextureRangeAPPLE(struct _glapi_table *disp, void (GLAPIENTRYP fn)(GLenum, GLsizei, const GLvoid *)) {
    SET_by_offset(disp, _gloffset_TextureRangeAPPLE, fn);
 }
 
diff -ur '--exclude-from=./mesa/src/aros/mesa.diff.excluded' -I '^.*\bRevision.*Date.*\b.*$' /data/deadwood/source/Mesa-7.11-staging/src/mesa/main/dlist.c ./mesa/src/mesa/main/dlist.c
--- /data/deadwood/source/Mesa-7.11-staging/src/mesa/main/dlist.c	2011-07-09 03:37:10.000000000 +0200
+++ ./mesa/src/mesa/main/dlist.c	2011-09-05 20:03:01.000000000 +0200
@@ -7541,6 +7541,7 @@
             CALL_Clear(ctx->Exec, (n[1].bf));
             break;
          case OPCODE_CLEAR_BUFFER_IV:
+#if 0 /* unused */
             {
                GLint value[4];
                value[0] = n[3].i;
@@ -7549,8 +7550,10 @@
                value[3] = n[6].i;
                /*CALL_ClearBufferiv(ctx->Exec, (n[1].e, n[2].i, value));*/
             }
+#endif
             break;
          case OPCODE_CLEAR_BUFFER_UIV:
+#if 0 /* unused */
             {
                GLuint value[4];
                value[0] = n[3].ui;
@@ -7559,8 +7562,10 @@
                value[3] = n[6].ui;
                /*CALL_ClearBufferiv(ctx->Exec, (n[1].e, n[2].i, value));*/
             }
+#endif
             break;
          case OPCODE_CLEAR_BUFFER_FV:
+#if 0 /* unused */
             {
                GLfloat value[4];
                value[0] = n[3].f;
@@ -7569,6 +7574,7 @@
                value[3] = n[6].f;
                /*CALL_ClearBufferfv(ctx->Exec, (n[1].e, n[2].i, value));*/
             }
+#endif
             break;
          case OPCODE_CLEAR_BUFFER_FI:
             /*CALL_ClearBufferfi(ctx->Exec, (n[1].e, n[2].i, n[3].f, n[4].i));*/
diff -ur '--exclude-from=./mesa/src/aros/mesa.diff.excluded' -I '^.*\bRevision.*Date.*\b.*$' /data/deadwood/source/Mesa-7.11-staging/src/mesa/main/imports.c ./mesa/src/mesa/main/imports.c
--- /data/deadwood/source/Mesa-7.11-staging/src/mesa/main/imports.c	2011-07-09 03:37:10.000000000 +0200
+++ ./mesa/src/mesa/main/imports.c	2011-09-05 20:03:01.000000000 +0200
@@ -453,14 +453,14 @@
 #endif
 }
 
-#ifndef __GNUC__
+#if !defined(__GNUC__) || defined(__AROS__)
 /**
  * Find the first bit set in a word.
  */
 int
 _mesa_ffs(int32_t i)
 {
-#if (defined(_WIN32) ) || defined(__IBMC__) || defined(__IBMCPP__)
+#if (defined(_WIN32) ) || defined(__IBMC__) || defined(__IBMCPP__) || defined(__AROS__)
    register int bit = 0;
    if (i != 0) {
       if ((i & 0xffff) == 0) {
@@ -513,7 +513,7 @@
 }
 #endif
 
-#if !defined(__GNUC__) ||\
+#if !defined(__GNUC__) || defined(__AROS__) ||\
    ((_GNUC__ == 3 && __GNUC_MINOR__ < 4) && __GNUC__ < 4)
 /**
  * Return number of bits set in given GLuint.
diff -ur '--exclude-from=./mesa/src/aros/mesa.diff.excluded' -I '^.*\bRevision.*Date.*\b.*$' /data/deadwood/source/Mesa-7.11-staging/src/mesa/main/imports.h ./mesa/src/mesa/main/imports.h
--- /data/deadwood/source/Mesa-7.11-staging/src/mesa/main/imports.h	2011-07-09 03:37:10.000000000 +0200
+++ ./mesa/src/mesa/main/imports.h	2011-09-05 20:03:01.000000000 +0200
@@ -560,7 +560,7 @@
 extern void
 _mesa_init_sqrt_table(void);
 
-#ifdef __GNUC__
+#if defined(__GNUC__) && !defined(__AROS__)
 
 #ifdef __MINGW32__
 #define ffs __builtin_ffs
diff -ur '--exclude-from=./mesa/src/aros/mesa.diff.excluded' -I '^.*\bRevision.*Date.*\b.*$' /data/deadwood/source/Mesa-7.11-staging/src/mesa/main/querymatrix.c ./mesa/src/mesa/main/querymatrix.c
--- /data/deadwood/source/Mesa-7.11-staging/src/mesa/main/querymatrix.c	2010-12-14 22:43:15.000000000 +0100
+++ ./mesa/src/mesa/main/querymatrix.c	2011-08-31 12:37:50.000000000 +0200
@@ -76,6 +76,10 @@
      (defined(__sun) && defined(__GNUC__))
 
 /* fpclassify is available. */
+#elif defined(__AROS__)
+
+#undef fpclassify
+#define fpclassify(x)   FP_NORMAL
 
 #elif !defined(_XOPEN_SOURCE) || _XOPEN_SOURCE < 600
 
diff -ur '--exclude-from=./mesa/src/aros/mesa.diff.excluded' -I '^.*\bRevision.*Date.*\b.*$' /data/deadwood/source/Mesa-7.11-staging/src/mesa/main/version.c ./mesa/src/mesa/main/version.c
--- /data/deadwood/source/Mesa-7.11-staging/src/mesa/main/version.c	2011-07-09 03:37:10.000000000 +0200
+++ ./mesa/src/mesa/main/version.c	2011-09-05 20:03:01.000000000 +0200
@@ -25,7 +25,9 @@
 #include "imports.h"
 #include "mtypes.h"
 #include "version.h"
+#ifdef MESA_GIT_SHA1
 #include "git_sha1.h"
+#endif
 
 
 
