diff -ruN boost_1_89_0/boost/assert/source_location.hpp boost_1_89_0.aros/boost/assert/source_location.hpp
--- boost_1_89_0/boost/assert/source_location.hpp	2025-08-06 19:49:08.000000000 +0100
+++ boost_1_89_0.aros/boost/assert/source_location.hpp	2026-07-28 12:54:50.146594725 +0100
@@ -76,7 +76,7 @@
 # pragma warning( disable: 4996 )
 #endif
 
-#if ( defined(_MSC_VER) && _MSC_VER < 1900 ) || ( defined(__MINGW32__) && !defined(__MINGW64_VERSION_MAJOR) )
+#if ( defined(_MSC_VER) && _MSC_VER < 1900 ) || ( defined(__MINGW32__) && !defined(__MINGW64_VERSION_MAJOR) ) || defined(__AROS__)
 # define BOOST_ASSERT_SNPRINTF(buffer, format, arg) std::sprintf(buffer, format, arg)
 #else
 # define BOOST_ASSERT_SNPRINTF(buffer, format, arg) std::snprintf(buffer, sizeof(buffer)/sizeof(buffer[0]), format, arg)
