diff -ruN libaom-3.12.1/aom_ports/ppc_cpudetect.c libaom-3.12.1.aros/aom_ports/ppc_cpudetect.c
--- libaom-3.12.1/aom_ports/ppc_cpudetect.c	2025-04-11 20:37:49.000000000 +0000
+++ libaom-3.12.1.aros/aom_ports/ppc_cpudetect.c	2025-08-02 16:55:10.420104471 +0000
@@ -12,14 +12,15 @@
 #include <fcntl.h>
 #include <unistd.h>
 #include <stdint.h>
-#include <asm/cputable.h>
-#include <linux/auxvec.h>
 
 #include "config/aom_config.h"
 
 #include "aom_ports/ppc.h"
 
 #if CONFIG_RUNTIME_CPU_DETECT
+#include <linux/auxvec.h>
+#include <asm/cputable.h>
+
 static int cpu_env_flags(int *flags) {
   char *env;
   env = getenv("AOM_SIMD_CAPS");
